// encodable/lib/parsers/scale/applyNice.js
"use strict";

exports.__esModule = true;
exports.default = applyNice;

var _d3Time = require("d3-time");

var _VegaLite = require("../../types/VegaLite");

const localTimeIntervals = {
  day: _d3Time.timeDay,
  hour: _d3Time.timeHour,
  minute: _d3Time.timeMinute,
  month: _d3Time.timeMonth,
  second: _d3Time.timeSecond,
  week: _d3Time.timeWeek,
  year: _d3Time.timeYear
};
const utcIntervals = {
  day: _d3Time.utcDay,
  hour: _d3Time.utcHour,
  minute: _d3Time.utcMinute,
  month: _d3Time.utcMonth,
  second: _d3Time.utcSecond,
  week: _d3Time.utcWeek,
  year: _d3Time.utcYear
};

function applyNice(config, scale) {
  if ('nice' in config && typeof config.nice !== 'undefined' && 'nice' in scale) {
    const {
      nice
    } = config;

    if (typeof nice === 'boolean') {
      if (nice) {
        scale.nice();
      }
    } else if (typeof nice === 'number') {
      scale.nice(nice);
    } else {
      const timeScale = scale;
      const {
        type
      } = config;

      if (typeof nice === 'string') {
        timeScale.nice(type === _VegaLite.ScaleType.UTC ? utcIntervals[nice] : localTimeIntervals[nice]);
      } else {
        const {
          interval,
          step
        } = nice;
        const parsedInterval = (type === _VegaLite.ScaleType.UTC ? utcIntervals[interval] : localTimeIntervals[interval]).every(step);

        if (parsedInterval !== null) {
          timeScale.nice(parsedInterval);
        }
      }
    }
  }
}
