// encodable/lib/parsers/scale/scaleCategories.js
"use strict";

exports.__esModule = true;
exports.allScaleTypesSet = exports.allScaleTypes = exports.timeScaleTypesSet = exports.timeScaleTypes = exports.discretizingScaleTypesSet = exports.discretizingScaleTypes = exports.discreteScaleTypesSet = exports.discreteScaleTypes = exports.continuousScaleTypesSet = exports.continuousScaleTypes = exports.discreteDomainScaleTypesSet = exports.discreteDomainScaleTypes = exports.continuousDomainScaleTypesSet = exports.continuousDomainScaleTypes = exports.continuousToDiscreteScaleTypesSet = exports.continuousToDiscreteScaleTypes = exports.continuousToContinuousScaleTypesSet = exports.continuousToContinuousScaleTypes = void 0;

var _VegaLite = require("../../types/VegaLite");

// Grouped by domain and range
const continuousToContinuousScaleTypes = [_VegaLite.ScaleType.LINEAR, _VegaLite.ScaleType.POW, _VegaLite.ScaleType.SQRT, _VegaLite.ScaleType.SYMLOG, _VegaLite.ScaleType.LOG, _VegaLite.ScaleType.TIME, _VegaLite.ScaleType.UTC];
exports.continuousToContinuousScaleTypes = continuousToContinuousScaleTypes;
const continuousToContinuousScaleTypesSet = new Set(continuousToContinuousScaleTypes);
exports.continuousToContinuousScaleTypesSet = continuousToContinuousScaleTypesSet;
const continuousToDiscreteScaleTypes = [_VegaLite.ScaleType.QUANTILE, _VegaLite.ScaleType.QUANTIZE, _VegaLite.ScaleType.THRESHOLD];
exports.continuousToDiscreteScaleTypes = continuousToDiscreteScaleTypes;
const continuousToDiscreteScaleTypesSet = new Set(continuousToDiscreteScaleTypes); // Grouped by Domain

exports.continuousToDiscreteScaleTypesSet = continuousToDiscreteScaleTypesSet;
const continuousDomainScaleTypes = continuousToContinuousScaleTypes.concat(continuousToDiscreteScaleTypes);
exports.continuousDomainScaleTypes = continuousDomainScaleTypes;
const continuousDomainScaleTypesSet = new Set(continuousDomainScaleTypes);
exports.continuousDomainScaleTypesSet = continuousDomainScaleTypesSet;
const discreteDomainScaleTypes = [_VegaLite.ScaleType.ORDINAL, _VegaLite.ScaleType.BIN_ORDINAL, _VegaLite.ScaleType.POINT, _VegaLite.ScaleType.BAND];
exports.discreteDomainScaleTypes = discreteDomainScaleTypes;
const discreteDomainScaleTypesSet = new Set(discreteDomainScaleTypes); // Three broad categories

exports.discreteDomainScaleTypesSet = discreteDomainScaleTypesSet;
const continuousScaleTypes = continuousToContinuousScaleTypes;
exports.continuousScaleTypes = continuousScaleTypes;
const continuousScaleTypesSet = continuousToContinuousScaleTypesSet;
exports.continuousScaleTypesSet = continuousScaleTypesSet;
const discreteScaleTypes = [_VegaLite.ScaleType.BAND, _VegaLite.ScaleType.POINT, _VegaLite.ScaleType.ORDINAL];
exports.discreteScaleTypes = discreteScaleTypes;
const discreteScaleTypesSet = new Set(discreteScaleTypes);
exports.discreteScaleTypesSet = discreteScaleTypesSet;
const discretizingScaleTypes = [_VegaLite.ScaleType.BIN_ORDINAL, _VegaLite.ScaleType.QUANTILE, _VegaLite.ScaleType.QUANTIZE, _VegaLite.ScaleType.THRESHOLD];
exports.discretizingScaleTypes = discretizingScaleTypes;
const discretizingScaleTypesSet = new Set(discretizingScaleTypes); // Others

exports.discretizingScaleTypesSet = discretizingScaleTypesSet;
const timeScaleTypes = [_VegaLite.ScaleType.TIME, _VegaLite.ScaleType.UTC];
exports.timeScaleTypes = timeScaleTypes;
const timeScaleTypesSet = new Set(timeScaleTypes);
exports.timeScaleTypesSet = timeScaleTypesSet;
const allScaleTypes = [_VegaLite.ScaleType.LINEAR, _VegaLite.ScaleType.LOG, _VegaLite.ScaleType.POW, _VegaLite.ScaleType.SQRT, _VegaLite.ScaleType.SYMLOG, _VegaLite.ScaleType.TIME, _VegaLite.ScaleType.UTC, _VegaLite.ScaleType.QUANTILE, _VegaLite.ScaleType.QUANTIZE, _VegaLite.ScaleType.THRESHOLD, _VegaLite.ScaleType.BIN_ORDINAL, _VegaLite.ScaleType.ORDINAL, _VegaLite.ScaleType.POINT, _VegaLite.ScaleType.BAND];
exports.allScaleTypes = allScaleTypes;
const allScaleTypesSet = new Set(allScaleTypes);
exports.allScaleTypesSet = allScaleTypesSet;
