// luma.gl/dist/es5/webgl-context/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createGLContext", {
  enumerable: true,
  get: function get() {
    return _context.createGLContext;
  }
});
Object.defineProperty(exports, "destroyGLContext", {
  enumerable: true,
  get: function get() {
    return _context.destroyGLContext;
  }
});
Object.defineProperty(exports, "resizeGLContext", {
  enumerable: true,
  get: function get() {
    return _context.resizeGLContext;
  }
});
Object.defineProperty(exports, "pollGLContext", {
  enumerable: true,
  get: function get() {
    return _context.pollGLContext;
  }
});
Object.defineProperty(exports, "withParameters", {
  enumerable: true,
  get: function get() {
    return _contextState.withParameters;
  }
});
Object.defineProperty(exports, "resetParameters", {
  enumerable: true,
  get: function get() {
    return _contextState.resetParameters;
  }
});
Object.defineProperty(exports, "getContextInfo", {
  enumerable: true,
  get: function get() {
    return _contextLimits.getContextInfo;
  }
});
Object.defineProperty(exports, "pageLoadPromise", {
  enumerable: true,
  get: function get() {
    return _createCanvas.pageLoadPromise;
  }
});
Object.defineProperty(exports, "getPageLoadPromise", {
  enumerable: true,
  get: function get() {
    return _createCanvas.getPageLoadPromise;
  }
});
Object.defineProperty(exports, "createCanvas", {
  enumerable: true,
  get: function get() {
    return _createCanvas.createCanvas;
  }
});
Object.defineProperty(exports, "getCanvas", {
  enumerable: true,
  get: function get() {
    return _createCanvas.getCanvas;
  }
});
Object.defineProperty(exports, "createHeadlessContext", {
  enumerable: true,
  get: function get() {
    return _createHeadlessContext.createHeadlessContext;
  }
});
Object.defineProperty(exports, "trackContextCreation", {
  enumerable: true,
  get: function get() {
    return _createBrowserContext.trackContextCreation;
  }
});
Object.defineProperty(exports, "createBrowserContext", {
  enumerable: true,
  get: function get() {
    return _createBrowserContext.createBrowserContext;
  }
});
Object.defineProperty(exports, "polyfillContext", {
  enumerable: true,
  get: function get() {
    return _polyfillContext.default;
  }
});
Object.defineProperty(exports, "trackContextState", {
  enumerable: true,
  get: function get() {
    return _trackContextState.default;
  }
});

var _context = require("./context");

var _contextState = require("./context-state");

var _contextLimits = require("./context-limits");

var _createCanvas = require("./create-canvas");

var _createHeadlessContext = require("./create-headless-context");

var _createBrowserContext = require("./create-browser-context");

var _polyfillContext = _interopRequireDefault(require("./polyfill-context"));

var _trackContextState = _interopRequireDefault(require("./track-context-state"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
