// @superset-ui/legacy-plugin-chart-parallel-coordinates/lib/ParallelCoordinates.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _d = _interopRequireDefault(require("d3"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _color = require("@superset-ui/color");

var _d2 = _interopRequireDefault(require("./vendor/parcoords/d3.parcoords"));

var _divgrid = _interopRequireDefault(require("./vendor/parcoords/divgrid"));

require("./vendor/parcoords/d3.parcoords.css");

require("./ParallelCoordinates.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable react/sort-prop-types */
const propTypes = {
  // Standard tabular data [{ fieldName1: value1, fieldName2: value2 }]
  data: _propTypes.default.arrayOf(_propTypes.default.object),
  width: _propTypes.default.number,
  height: _propTypes.default.number,
  colorMetric: _propTypes.default.string,
  includeSeries: _propTypes.default.bool,
  linearColorScheme: _propTypes.default.string,
  metrics: _propTypes.default.arrayOf(_propTypes.default.string),
  series: _propTypes.default.string,
  showDatatable: _propTypes.default.bool
};

function ParallelCoordinates(element, props) {
  const {
    data,
    width,
    height,
    colorMetric,
    includeSeries,
    linearColorScheme,
    metrics,
    series,
    showDatatable
  } = props;
  const cols = includeSeries ? [series].concat(metrics) : metrics;
  const ttypes = {};
  ttypes[series] = 'string';
  metrics.forEach(v => {
    ttypes[v] = 'number';
  });
  const colorScale = colorMetric ? (0, _color.getSequentialSchemeRegistry)().get(linearColorScheme).createLinearScale(_d.default.extent(data, d => d[colorMetric])) : () => 'grey';

  const color = d => colorScale(d[colorMetric]);

  const container = _d.default.select(element).classed('superset-legacy-chart-parallel-coordinates', true);

  container.selectAll('*').remove();
  const effHeight = showDatatable ? height / 2 : height;
  const div = container.append('div').style('height', effHeight + "px").classed('parcoords', true);
  const chart = (0, _d2.default)()(div.node()).width(width).color(color).alpha(0.5).composite('darken').height(effHeight).data(data).dimensions(cols).types(ttypes).render().createAxes().shadows().reorderable().brushMode('1D-axes');

  if (showDatatable) {
    // create data table, row hover highlighting
    const grid = (0, _divgrid.default)();
    container.append('div').style('height', effHeight + "px").datum(data).call(grid).classed('parcoords grid', true).selectAll('.row').on({
      mouseover(d) {
        chart.highlight([d]);
      },

      mouseout: chart.unhighlight
    }); // update data table on brush event

    chart.on('brush', d => {
      _d.default.select('.grid').datum(d).call(grid).selectAll('.row').on({
        mouseover(dd) {
          chart.highlight([dd]);
        },

        mouseout: chart.unhighlight
      });
    });
  }
}

ParallelCoordinates.displayName = 'ParallelCoordinates';
ParallelCoordinates.propTypes = propTypes;
var _default = ParallelCoordinates;
exports.default = _default;
