// @superset-ui/legacy-plugin-chart-histogram/esm/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _translation = require("@superset-ui/translation");

var _validator = require("@superset-ui/validator");

var _chartControls = require("@superset-ui/chart-controls");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const {
  columnChoices
} = _chartControls.internalSharedControls;
var _default = {
  controlPanelSections: [{
    label: (0, _translation.t)('Query'),
    expanded: true,
    controlSetRows: [[{
      name: 'all_columns_x',
      config: {
        type: 'SelectControl',
        label: (0, _translation.t)('Numeric Columns'),
        default: null,
        description: (0, _translation.t)('Select the numeric columns to draw the histogram'),
        mapStateToProps: state => ({
          choices: columnChoices(state.datasource)
        }),
        multi: true,
        validators: [_validator.validateNonEmpty]
      }
    }], ['adhoc_filters'], ['row_limit'], ['groupby']]
  }, {
    label: (0, _translation.t)('Chart Options'),
    expanded: true,
    controlSetRows: [['color_scheme', 'label_colors'], [{
      name: 'link_length',
      config: {
        type: 'SelectControl',
        renderTrigger: true,
        freeForm: true,
        label: (0, _translation.t)('No of Bins'),
        default: 5,
        choices: (0, _chartControls.formatSelectOptions)(['10', '25', '50', '75', '100', '150', '200', '250']),
        description: (0, _translation.t)('Select the number of bins for the histogram')
      }
    }], [{
      name: 'x_axis_label',
      config: {
        type: 'TextControl',
        label: (0, _translation.t)('X Axis Label'),
        renderTrigger: true,
        default: ''
      }
    }, {
      name: 'y_axis_label',
      config: {
        type: 'TextControl',
        label: (0, _translation.t)('Y Axis Label'),
        renderTrigger: true,
        default: ''
      }
    }], [{
      name: 'normalized',
      config: {
        type: 'CheckboxControl',
        label: (0, _translation.t)('Normalized'),
        renderTrigger: true,
        description: (0, _translation.t)('Whether to normalize the histogram'),
        default: false
      }
    }]]
  }]
};
exports.default = _default;
