// eslint-plugin-jsx-a11y/lib/util/implicitRoles/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _a = require('./a');

var _a2 = _interopRequireDefault(_a);

var _area = require('./area');

var _area2 = _interopRequireDefault(_area);

var _article = require('./article');

var _article2 = _interopRequireDefault(_article);

var _aside = require('./aside');

var _aside2 = _interopRequireDefault(_aside);

var _body = require('./body');

var _body2 = _interopRequireDefault(_body);

var _button = require('./button');

var _button2 = _interopRequireDefault(_button);

var _datalist = require('./datalist');

var _datalist2 = _interopRequireDefault(_datalist);

var _details = require('./details');

var _details2 = _interopRequireDefault(_details);

var _dialog = require('./dialog');

var _dialog2 = _interopRequireDefault(_dialog);

var _dl = require('./dl');

var _dl2 = _interopRequireDefault(_dl);

var _form = require('./form');

var _form2 = _interopRequireDefault(_form);

var _h = require('./h1');

var _h2 = _interopRequireDefault(_h);

var _h3 = require('./h2');

var _h4 = _interopRequireDefault(_h3);

var _h5 = require('./h3');

var _h6 = _interopRequireDefault(_h5);

var _h7 = require('./h4');

var _h8 = _interopRequireDefault(_h7);

var _h9 = require('./h5');

var _h10 = _interopRequireDefault(_h9);

var _h11 = require('./h6');

var _h12 = _interopRequireDefault(_h11);

var _hr = require('./hr');

var _hr2 = _interopRequireDefault(_hr);

var _img = require('./img');

var _img2 = _interopRequireDefault(_img);

var _input = require('./input');

var _input2 = _interopRequireDefault(_input);

var _li = require('./li');

var _li2 = _interopRequireDefault(_li);

var _link = require('./link');

var _link2 = _interopRequireDefault(_link);

var _menu = require('./menu');

var _menu2 = _interopRequireDefault(_menu);

var _menuitem = require('./menuitem');

var _menuitem2 = _interopRequireDefault(_menuitem);

var _meter = require('./meter');

var _meter2 = _interopRequireDefault(_meter);

var _nav = require('./nav');

var _nav2 = _interopRequireDefault(_nav);

var _ol = require('./ol');

var _ol2 = _interopRequireDefault(_ol);

var _option = require('./option');

var _option2 = _interopRequireDefault(_option);

var _output = require('./output');

var _output2 = _interopRequireDefault(_output);

var _progress = require('./progress');

var _progress2 = _interopRequireDefault(_progress);

var _section = require('./section');

var _section2 = _interopRequireDefault(_section);

var _select = require('./select');

var _select2 = _interopRequireDefault(_select);

var _tbody = require('./tbody');

var _tbody2 = _interopRequireDefault(_tbody);

var _textarea = require('./textarea');

var _textarea2 = _interopRequireDefault(_textarea);

var _tfoot = require('./tfoot');

var _tfoot2 = _interopRequireDefault(_tfoot);

var _thead = require('./thead');

var _thead2 = _interopRequireDefault(_thead);

var _ul = require('./ul');

var _ul2 = _interopRequireDefault(_ul);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
  a: _a2.default,
  area: _area2.default,
  article: _article2.default,
  aside: _aside2.default,
  body: _body2.default,
  button: _button2.default,
  datalist: _datalist2.default,
  details: _details2.default,
  dialog: _dialog2.default,
  dl: _dl2.default,
  form: _form2.default,
  h1: _h2.default,
  h2: _h4.default,
  h3: _h6.default,
  h4: _h8.default,
  h5: _h10.default,
  h6: _h12.default,
  hr: _hr2.default,
  img: _img2.default,
  input: _input2.default,
  li: _li2.default,
  link: _link2.default,
  menu: _menu2.default,
  menuitem: _menuitem2.default,
  meter: _meter2.default,
  nav: _nav2.default,
  ol: _ol2.default,
  option: _option2.default,
  output: _output2.default,
  progress: _progress2.default,
  section: _section2.default,
  select: _select2.default,
  tbody: _tbody2.default,
  textarea: _textarea2.default,
  tfoot: _tfoot2.default,
  thead: _thead2.default,
  ul: _ul2.default
};
