// @data-ui/xy-chart/lib/series/AreaSeries.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _shape = require("@vx/shape");

var _theme = require("@data-ui/theme");

var _shared = require("@data-ui/shared");

var _group = require("@vx/group");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _propShapes = require("../utils/propShapes");

var _chartUtils = require("../utils/chartUtils");

var _findClosestDatum = _interopRequireDefault(require("../utils/findClosestDatum"));

var _interpolatorLookup = _interopRequireDefault(require("../utils/interpolatorLookup"));

var _sharedSeriesProps = _interopRequireDefault(require("../utils/sharedSeriesProps"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var propTypes = _extends({}, _sharedSeriesProps.default, {
  data: _propShapes.areaSeriesDataShape.isRequired,
  fill: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  fillOpacity: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.number]),
  interpolation: _propShapes.interpolationShape,
  stroke: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  strokeDasharray: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  strokeWidth: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.number]),
  strokeLinecap: _propTypes.default.oneOf(['butt', 'square', 'round', 'inherit'])
});

var defaultProps = {
  interpolation: 'monotoneX',
  stroke: _theme.color.default,
  strokeWidth: 3,
  strokeDasharray: null,
  strokeLinecap: 'round',
  fill: _theme.color.default,
  fillOpacity: 0.3
};

var x = function x(d) {
  return d && d.x;
};

var getY = function getY(d) {
  return d && d.y;
};

var getY0 = function getY0(d) {
  return d && d.y0;
};

var getY1 = function getY1(d) {
  return d && d.y1;
};

var definedClosed = function definedClosed(d) {
  return (0, _chartUtils.isDefined)(getY(d));
};

var definedOpen = function definedOpen(d) {
  return (0, _chartUtils.isDefined)(getY0(d)) && (0, _chartUtils.isDefined)(getY1(d));
};

var noEventsStyles = {
  pointerEvents: 'none'
};

var AreaSeries =
/*#__PURE__*/
function (_React$PureComponent) {
  _inheritsLoose(AreaSeries, _React$PureComponent);

  function AreaSeries() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = AreaSeries.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        data = _this$props.data,
        disableMouseEvents = _this$props.disableMouseEvents,
        xScale = _this$props.xScale,
        yScale = _this$props.yScale,
        margin = _this$props.margin,
        stroke = _this$props.stroke,
        strokeWidth = _this$props.strokeWidth,
        strokeDasharray = _this$props.strokeDasharray,
        strokeLinecap = _this$props.strokeLinecap,
        fill = _this$props.fill,
        fillOpacity = _this$props.fillOpacity,
        interpolation = _this$props.interpolation,
        onClick = _this$props.onClick,
        onMouseMove = _this$props.onMouseMove,
        onMouseLeave = _this$props.onMouseLeave;
    if (!xScale || !yScale) return null;
    var datum0 = data[0] || {};
    var isClosed = !definedOpen(datum0);
    var yMin = yScale.domain()[0];
    var y0 = isClosed ? function () {
      return yMin;
    } : getY0;
    var y1 = isClosed ? getY : getY1;
    var defined = isClosed ? definedClosed : definedOpen;
    var strokeDasharrayValue = (0, _chartUtils.callOrValue)(strokeDasharray, data);
    var strokeValue = (0, _chartUtils.callOrValue)(stroke, data);
    var strokeWidthValue = (0, _chartUtils.callOrValue)(strokeWidth, data);
    var fillValue = (0, _chartUtils.callOrValue)(fill, data);
    var curve = _interpolatorLookup.default[interpolation] || _interpolatorLookup.default.monotoneX;
    return _react.default.createElement(_group.Group, {
      style: disableMouseEvents ? noEventsStyles : null,
      onClick: disableMouseEvents ? null : onClick && function (event) {
        var d = (0, _findClosestDatum.default)({
          data: data,
          getX: x,
          event: event,
          xScale: xScale,
          marginLeft: margin.left
        });
        onClick({
          event: event,
          data: data,
          datum: d,
          color: fillValue
        });
      },
      onMouseMove: disableMouseEvents ? null : onMouseMove && function (event) {
        var d = (0, _findClosestDatum.default)({
          data: data,
          getX: x,
          event: event,
          xScale: xScale,
          marginLeft: margin.left
        });
        onMouseMove({
          event: event,
          data: data,
          datum: d,
          color: fillValue
        });
      },
      onMouseLeave: disableMouseEvents ? null : onMouseLeave
    }, _react.default.createElement(_shape.Area, {
      data: data,
      x: x,
      y0: y0,
      y1: y1,
      xScale: xScale,
      yScale: yScale,
      fill: fillValue,
      fillOpacity: (0, _chartUtils.callOrValue)(fillOpacity, data),
      stroke: "transparent",
      strokeWidth: strokeWidthValue,
      curve: curve,
      defined: defined
    }), strokeWidthValue > 0 && !isClosed && _react.default.createElement(_shape.LinePath, {
      data: data,
      x: x,
      y: y0,
      xScale: xScale,
      yScale: yScale,
      stroke: strokeValue,
      strokeWidth: strokeWidthValue,
      strokeDasharray: strokeDasharrayValue,
      strokeLinecap: strokeLinecap,
      curve: curve,
      glyph: null,
      defined: defined
    }), _react.default.createElement(_shape.LinePath, {
      data: data,
      x: x,
      y: y1,
      xScale: xScale,
      yScale: yScale,
      stroke: strokeValue,
      strokeWidth: strokeWidthValue,
      strokeDasharray: strokeDasharrayValue,
      strokeLinecap: strokeLinecap,
      curve: curve,
      defined: defined,
      glyph: function glyph(d, i) {
        return _react.default.createElement(_shared.FocusBlurHandler, {
          key: "areapoint-" + i,
          onBlur: disableMouseEvents ? null : onMouseLeave,
          onFocus: disableMouseEvents ? null : function (event) {
            onMouseMove({
              event: event,
              data: data,
              datum: d,
              color: strokeValue,
              index: i
            });
          }
        });
      }
    }));
  };

  return AreaSeries;
}(_react.default.PureComponent);

exports.default = AreaSeries;
AreaSeries.propTypes = propTypes;
AreaSeries.defaultProps = defaultProps;
AreaSeries.displayName = 'AreaSeries';
