// @superset-ui/legacy-plugin-chart-markup/esm/Markup.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _style = _interopRequireDefault(require("@superset-ui/style"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _templateObject() {
  const data = _taggedTemplateLiteralLoose(["\n  .markup.slice_container {\n    margin: 10px;\n  }\n  .separator {\n    background-color: transparent !important;\n  }\n  .separator hr {\n    border: 0;\n    height: 1px;\n    background-image: linear-gradient(\n      to right,\n      rgba(0, 0, 0, 1),\n      rgba(0, 0, 0, 1),\n      rgba(0, 0, 0, 1),\n      rgba(0, 0, 0, 0)\n    );\n  }\n  .separator .chart-header {\n    border: none !important;\n  }\n"]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteralLoose(strings, raw) { if (!raw) { raw = strings.slice(0); } strings.raw = raw; return strings; }

const propTypes = {
  className: _propTypes.default.string,
  cssFiles: _propTypes.default.arrayOf(_propTypes.default.string),
  height: _propTypes.default.number.isRequired,
  html: _propTypes.default.string,
  isSeparator: _propTypes.default.bool
};
const defaultProps = {
  className: '',
  cssFiles: [],
  html: '',
  isSeparator: false
};
const CONTAINER_STYLE = {
  overflow: 'auto',
  position: 'relative'
};

class Markup extends _react.default.PureComponent {
  render() {
    const {
      className,
      height,
      isSeparator,
      html,
      cssFiles
    } = this.props;
    return /*#__PURE__*/_react.default.createElement("div", {
      className: className,
      style: CONTAINER_STYLE
    }, /*#__PURE__*/_react.default.createElement("iframe", {
      title: "superset-markup",
      frameBorder: 0,
      height: isSeparator ? height - 20 : height,
      sandbox: "allow-forms allow-popups allow-same-origin allow-scripts allow-top-navigation",
      srcDoc: "\n            <html>\n              <head>\n                " + cssFiles.map(href => "<link rel=\"stylesheet\" type=\"text/css\" href=\"" + href + "\" />") + "\n              </head>\n              <body style=\"background-color: transparent;\">\n                " + html + "\n              </body>\n            </html>"
    }));
  }

}

Markup.propTypes = propTypes;
Markup.defaultProps = defaultProps;

var _default = (0, _style.default)(Markup)(_templateObject());

exports.default = _default;
