// @superset-ui/legacy-preset-chart-nvd3/lib/Bubble/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _translation = require("@superset-ui/translation");

var _chartControls = require("@superset-ui/chart-controls");

var _NVD3Controls = require("../NVD3Controls");

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _default = {
  label: (0, _translation.t)('Bubble Chart'),
  controlPanelSections: [{
    label: (0, _translation.t)('Query'),
    expanded: true,
    controlSetRows: [['series', 'entity'], ['x'], ['y'], ['adhoc_filters'], ['size'], [{
      name: 'max_bubble_size',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _translation.t)('Max Bubble Size'),
        default: '25',
        choices: (0, _chartControls.formatSelectOptions)(['5', '10', '15', '25', '50', '75', '100'])
      }
    }], ['limit', null]]
  }, {
    label: (0, _translation.t)('Chart Options'),
    expanded: true,
    controlSetRows: [['color_scheme', 'label_colors'], [_NVD3Controls.showLegend, null]]
  }, {
    label: (0, _translation.t)('X Axis'),
    expanded: true,
    controlSetRows: [[_NVD3Controls.xAxisLabel, _NVD3Controls.leftMargin], [{
      name: _NVD3Controls.xAxisFormat.name,
      config: _extends({}, _NVD3Controls.xAxisFormat.config, {
        default: 'SMART_NUMBER',
        choices: _chartControls.D3_FORMAT_OPTIONS
      })
    }, _NVD3Controls.xTicksLayout], [{
      name: 'x_log_scale',
      config: {
        type: 'CheckboxControl',
        label: (0, _translation.t)('X Log Scale'),
        default: false,
        renderTrigger: true,
        description: (0, _translation.t)('Use a log scale for the X-axis')
      }
    }, _NVD3Controls.xAxisShowMinmax]]
  }, {
    label: (0, _translation.t)('Y Axis'),
    expanded: true,
    controlSetRows: [[_NVD3Controls.yAxisLabel, _NVD3Controls.bottomMargin], ['y_axis_format', null], [_NVD3Controls.yLogScale, _NVD3Controls.yAxisShowMinmax]]
  }],
  controlOverrides: {
    color_scheme: {
      renderTrigger: false
    }
  }
};
exports.default = _default;
