// jsdom/lib/jsdom/living/generated/HTMLMetaElement.js
"use strict";

const conversions = require("webidl-conversions");
const utils = require("./utils.js");
const HTMLElement = require("./HTMLElement.js");
const impl = utils.implSymbol;

function HTMLMetaElement() {
  throw new TypeError("Illegal constructor");
}
HTMLMetaElement.prototype = Object.create(HTMLElement.interface.prototype);
HTMLMetaElement.prototype.constructor = HTMLMetaElement;


HTMLMetaElement.prototype.toString = function () {
  if (this === HTMLMetaElement.prototype) {
    return "[object HTMLMetaElementPrototype]";
  }
  return HTMLElement.interface.prototype.toString.call(this);
};
Object.defineProperty(HTMLMetaElement.prototype, "name", {
  get() {
    const value = this.getAttribute("name");
    return value === null ? "" : value;
  },
  set(V) {
    V = conversions["DOMString"](V);
    this.setAttribute("name", V);
  },
  enumerable: true,
  configurable: true
});

Object.defineProperty(HTMLMetaElement.prototype, "httpEquiv", {
  get() {
    const value = this.getAttribute("http-equiv");
    return value === null ? "" : value;
  },
  set(V) {
    V = conversions["DOMString"](V);
    this.setAttribute("http-equiv", V);
  },
  enumerable: true,
  configurable: true
});

Object.defineProperty(HTMLMetaElement.prototype, "content", {
  get() {
    const value = this.getAttribute("content");
    return value === null ? "" : value;
  },
  set(V) {
    V = conversions["DOMString"](V);
    this.setAttribute("content", V);
  },
  enumerable: true,
  configurable: true
});

Object.defineProperty(HTMLMetaElement.prototype, "scheme", {
  get() {
    const value = this.getAttribute("scheme");
    return value === null ? "" : value;
  },
  set(V) {
    V = conversions["DOMString"](V);
    this.setAttribute("scheme", V);
  },
  enumerable: true,
  configurable: true
});


const iface = {
  mixedInto: [],
  is(obj) {
    if (obj) {
      if (obj[impl] instanceof Impl.implementation) {
        return true;
      }
      for (let i = 0; i < module.exports.mixedInto.length; ++i) {
        if (obj instanceof module.exports.mixedInto[i]) {
          return true;
        }
      }
    }
    return false;
  },
  isImpl(obj) {
    if (obj) {
      if (obj instanceof Impl.implementation) {
        return true;
      }

      const wrapper = utils.wrapperForImpl(obj);
      for (let i = 0; i < module.exports.mixedInto.length; ++i) {
        if (wrapper instanceof module.exports.mixedInto[i]) {
          return true;
        }
      }
    }
    return false;
  },
  create(constructorArgs, privateData) {
    let obj = Object.create(HTMLMetaElement.prototype);
    this.setup(obj, constructorArgs, privateData);
    return obj;
  },
  createImpl(constructorArgs, privateData) {
    let obj = Object.create(HTMLMetaElement.prototype);
    this.setup(obj, constructorArgs, privateData);
    return utils.implForWrapper(obj);
  },
  _internalSetup(obj) {
    HTMLElement._internalSetup(obj);

  },
  setup(obj, constructorArgs, privateData) {
    if (!privateData) privateData = {};
    privateData.wrapper = obj;

    this._internalSetup(obj);

    obj[impl] = new Impl.implementation(constructorArgs, privateData);
    obj[impl][utils.wrapperSymbol] = obj;
  },
  interface: HTMLMetaElement,
  expose: {
    Window: { HTMLMetaElement: HTMLMetaElement }
  }
};
module.exports = iface;

const Impl = require("../nodes/HTMLMetaElement-impl.js");
