// @superset-ui/preset-chart-xy/lib/index.js
"use strict";

exports.__esModule = true;
exports.LegacyLineChartPlugin = exports.LineChartPlugin = exports.LegacyScatterPlotChartPlugin = exports.ScatterPlotChartPlugin = exports.LegacyBoxPlotChartPlugin = exports.BoxPlotChartPlugin = void 0;

var _BoxPlot = _interopRequireDefault(require("./BoxPlot"));

exports.BoxPlotChartPlugin = _BoxPlot.default;

var _legacy = _interopRequireDefault(require("./BoxPlot/legacy"));

exports.LegacyBoxPlotChartPlugin = _legacy.default;

var _ScatterPlot = _interopRequireDefault(require("./ScatterPlot"));

exports.ScatterPlotChartPlugin = _ScatterPlot.default;

var _legacy2 = _interopRequireDefault(require("./ScatterPlot/legacy"));

exports.LegacyScatterPlotChartPlugin = _legacy2.default;

var _Line = _interopRequireDefault(require("./Line"));

exports.LineChartPlugin = _Line.default;

var _legacy3 = _interopRequireDefault(require("./Line/legacy"));

exports.LegacyLineChartPlugin = _legacy3.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
