// @vx/shape/lib/shapes/Pie.js
"use strict";

exports.__esModule = true;
exports.default = Pie;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _group = require("@vx/group");

var _d3Shape = require("d3-shape");

var _setNumberOrNumberAccessor = _interopRequireDefault(require("../util/setNumberOrNumberAccessor"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function Pie(_ref) {
  var className = _ref.className,
      top = _ref.top,
      left = _ref.left,
      _ref$data = _ref.data,
      data = _ref$data === void 0 ? [] : _ref$data,
      centroid = _ref.centroid,
      _ref$innerRadius = _ref.innerRadius,
      innerRadius = _ref$innerRadius === void 0 ? 0 : _ref$innerRadius,
      outerRadius = _ref.outerRadius,
      cornerRadius = _ref.cornerRadius,
      startAngle = _ref.startAngle,
      endAngle = _ref.endAngle,
      padAngle = _ref.padAngle,
      padRadius = _ref.padRadius,
      pieSort = _ref.pieSort,
      pieSortValues = _ref.pieSortValues,
      pieValue = _ref.pieValue,
      children = _ref.children,
      restProps = _objectWithoutPropertiesLoose(_ref, ["className", "top", "left", "data", "centroid", "innerRadius", "outerRadius", "cornerRadius", "startAngle", "endAngle", "padAngle", "padRadius", "pieSort", "pieSortValues", "pieValue", "children"]);

  var path = (0, _d3Shape.arc)();
  if (innerRadius != null) (0, _setNumberOrNumberAccessor.default)(path.innerRadius, innerRadius);
  if (outerRadius != null) (0, _setNumberOrNumberAccessor.default)(path.outerRadius, outerRadius);
  if (cornerRadius != null) (0, _setNumberOrNumberAccessor.default)(path.cornerRadius, cornerRadius);
  if (padRadius != null) (0, _setNumberOrNumberAccessor.default)(path.padRadius, padRadius);
  var pie = (0, _d3Shape.pie)(); // ts can't distinguish between these method overloads

  if (pieSort === null) pie.sort(pieSort);else if (pieSort != null) pie.sort(pieSort);
  if (pieSortValues === null) pie.sortValues(pieSortValues);else if (pieSortValues != null) pie.sortValues(pieSortValues);
  if (pieValue != null) pie.value(pieValue);
  if (padAngle != null) (0, _setNumberOrNumberAccessor.default)(pie.padAngle, padAngle);
  if (startAngle != null) (0, _setNumberOrNumberAccessor.default)(pie.startAngle, startAngle);
  if (endAngle != null) (0, _setNumberOrNumberAccessor.default)(pie.endAngle, endAngle);
  var arcs = pie(data);
  if (children) return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, children({
    arcs: arcs,
    path: path,
    pie: pie
  }));
  return /*#__PURE__*/_react.default.createElement(_group.Group, {
    className: "vx-pie-arcs-group",
    top: top,
    left: left
  }, arcs.map(function (arc, i) {
    return /*#__PURE__*/_react.default.createElement("g", {
      key: "pie-arc-" + i
    }, /*#__PURE__*/_react.default.createElement("path", _extends({
      className: (0, _classnames.default)('vx-pie-arc', className),
      d: path(arc) || ''
    }, restProps)), centroid && centroid(path.centroid(arc), arc));
  }));
}

Pie.propTypes = {
  className: _propTypes.default.string,
  top: _propTypes.default.number,
  left: _propTypes.default.number,
  data: _propTypes.default.array,
  pieSort: _propTypes.default.oneOfType([_propTypes.default.oneOf([null]), _propTypes.default.func]),
  pieSortValues: _propTypes.default.oneOfType([_propTypes.default.oneOf([null]), _propTypes.default.func]),
  centroid: _propTypes.default.func,
  children: _propTypes.default.func
};
