// @superset-ui/chart-controls/esm/sections.js
"use strict";

exports.__esModule = true;
exports.annotations = exports.sqlaTimeSeries = exports.colorScheme = exports.datasourceAndVizType = exports.druidTimeSeries = void 0;

var _translation = require("@superset-ui/translation");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
// A few standard controls sections that are used internally.
// Not recommended for use in third-party plugins.
const druidTimeSeries = {
  label: (0, _translation.t)('Time'),
  expanded: true,
  description: (0, _translation.t)('Time related form attributes'),
  controlSetRows: [['time_range']]
};
exports.druidTimeSeries = druidTimeSeries;
const datasourceAndVizType = {
  label: (0, _translation.t)('Datasource & Chart Type'),
  expanded: true,
  controlSetRows: [['datasource'], ['viz_type'], [{
    name: 'slice_id',
    config: {
      type: 'HiddenControl',
      label: (0, _translation.t)('Chart ID'),
      hidden: true,
      description: (0, _translation.t)('The id of the active chart')
    }
  }, {
    name: 'cache_timeout',
    config: {
      type: 'HiddenControl',
      label: (0, _translation.t)('Cache Timeout (seconds)'),
      hidden: true,
      description: (0, _translation.t)('The number of seconds before expiring the cache')
    }
  }, {
    name: 'url_params',
    config: {
      type: 'HiddenControl',
      label: (0, _translation.t)('URL Parameters'),
      hidden: true,
      description: (0, _translation.t)('Extra parameters for use in jinja templated queries')
    }
  }, {
    name: 'time_range_endpoints',
    config: {
      type: 'HiddenControl',
      label: (0, _translation.t)('Time range endpoints'),
      hidden: true,
      description: (0, _translation.t)('Time range endpoints (SIP-15)')
    }
  }]]
};
exports.datasourceAndVizType = datasourceAndVizType;
const colorScheme = {
  label: (0, _translation.t)('Color Scheme'),
  controlSetRows: [['color_scheme', 'label_colors']]
};
exports.colorScheme = colorScheme;
const sqlaTimeSeries = {
  label: (0, _translation.t)('Time'),
  description: (0, _translation.t)('Time related form attributes'),
  expanded: true,
  controlSetRows: [['granularity_sqla'], ['time_range']]
};
exports.sqlaTimeSeries = sqlaTimeSeries;
const annotations = {
  label: (0, _translation.t)('Annotations and Layers'),
  tabOverride: 'data',
  expanded: true,
  controlSetRows: [[{
    name: 'annotation_layers',
    config: {
      type: 'AnnotationLayerControl',
      label: '',
      default: [],
      description: 'Annotation Layers',
      renderTrigger: true,
      tabOverride: 'data'
    }
  }]]
};
exports.annotations = annotations;
