// @data-ui/histogram/lib/utils/propShapes.js
"use strict";

exports.__esModule = true;
exports.themeShape = exports.gridStylesShape = exports.tickStylesShape = exports.axisStylesShape = exports.binnedDataShape = exports.categoricalBinnedDatumShape = exports.numericBinnedDatumShape = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var numericBinnedDatumShape = _propTypes.default.shape({
  id: _propTypes.default.string.isRequired,
  bin0: _propTypes.default.number.isRequired,
  bin1: _propTypes.default.number.isRequired,
  count: _propTypes.default.number.isRequired,
  cumulative: _propTypes.default.number,
  density: _propTypes.default.number,
  cumulativeDensity: _propTypes.default.number
});

exports.numericBinnedDatumShape = numericBinnedDatumShape;

var categoricalBinnedDatumShape = _propTypes.default.shape({
  id: _propTypes.default.string.isRequired,
  bin: _propTypes.default.string.isRequired,
  count: _propTypes.default.number.isRequired,
  cumulative: _propTypes.default.number,
  density: _propTypes.default.number,
  cumulativeDensity: _propTypes.default.number
});

exports.categoricalBinnedDatumShape = categoricalBinnedDatumShape;

var binnedDataShape = _propTypes.default.arrayOf(_propTypes.default.oneOfType([numericBinnedDatumShape, categoricalBinnedDatumShape])); // styles ---------------------------------------------------------------------


exports.binnedDataShape = binnedDataShape;

var axisStylesShape = _propTypes.default.shape({
  stroke: _propTypes.default.string,
  strokeWidth: _propTypes.default.number,
  label: _propTypes.default.shape({
    left: _propTypes.default.object,
    right: _propTypes.default.object,
    bottom: _propTypes.default.object,
    top: _propTypes.default.object
  })
});

exports.axisStylesShape = axisStylesShape;

var tickStylesShape = _propTypes.default.shape({
  stroke: _propTypes.default.string,
  tickLength: _propTypes.default.number,
  label: _propTypes.default.shape({
    left: _propTypes.default.object,
    right: _propTypes.default.object,
    bottom: _propTypes.default.object,
    top: _propTypes.default.object
  })
});

exports.tickStylesShape = tickStylesShape;

var gridStylesShape = _propTypes.default.shape({
  stroke: _propTypes.default.string,
  strokeWidth: _propTypes.default.number
});

exports.gridStylesShape = gridStylesShape;

var themeShape = _propTypes.default.shape({
  gridStyles: gridStylesShape,
  xAxisStyles: axisStylesShape,
  xTickStyles: tickStylesShape,
  yAxisStyles: axisStylesShape,
  yTickStyles: tickStylesShape
});

exports.themeShape = themeShape;
