// encodable/esm/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  Encoder: true,
  ChannelEncoder: true,
  createEncoderFactory: true,
  completeChannelDef: true,
  createScaleFromScaleConfig: true,
  mergeEncoding: true,
  isFieldDef: true,
  isNonValueDef: true,
  isTypedFieldDef: true,
  isValueDef: true,
  isScaleFieldDef: true,
  isPositionFieldDef: true,
  isCompleteFieldDef: true,
  isCompleteValueDef: true
};
exports.isCompleteValueDef = exports.isCompleteFieldDef = exports.isPositionFieldDef = exports.isScaleFieldDef = exports.isValueDef = exports.isTypedFieldDef = exports.isNonValueDef = exports.isFieldDef = exports.mergeEncoding = exports.createScaleFromScaleConfig = exports.completeChannelDef = exports.createEncoderFactory = exports.ChannelEncoder = exports.Encoder = void 0;

var _Encoder = _interopRequireDefault(require("./encoders/Encoder"));

exports.Encoder = _Encoder.default;

var _ChannelEncoder = _interopRequireDefault(require("./encoders/ChannelEncoder"));

exports.ChannelEncoder = _ChannelEncoder.default;

var _createEncoderFactory = _interopRequireDefault(require("./encoders/createEncoderFactory"));

exports.createEncoderFactory = _createEncoderFactory.default;

var _completeChannelDef = _interopRequireDefault(require("./fillers/completeChannelDef"));

exports.completeChannelDef = _completeChannelDef.default;

var _createScaleFromScaleConfig = _interopRequireDefault(require("./parsers/scale/createScaleFromScaleConfig"));

exports.createScaleFromScaleConfig = _createScaleFromScaleConfig.default;

var _mergeEncoding = _interopRequireDefault(require("./utils/mergeEncoding"));

exports.mergeEncoding = _mergeEncoding.default;

var _ChannelDef = require("./typeGuards/ChannelDef");

exports.isFieldDef = _ChannelDef.isFieldDef;
exports.isNonValueDef = _ChannelDef.isNonValueDef;
exports.isTypedFieldDef = _ChannelDef.isTypedFieldDef;
exports.isValueDef = _ChannelDef.isValueDef;
exports.isScaleFieldDef = _ChannelDef.isScaleFieldDef;
exports.isPositionFieldDef = _ChannelDef.isPositionFieldDef;

var _CompleteChannelDef = require("./typeGuards/CompleteChannelDef");

exports.isCompleteFieldDef = _CompleteChannelDef.isCompleteFieldDef;
exports.isCompleteValueDef = _CompleteChannelDef.isCompleteValueDef;

var _Axis = require("./types/Axis");

Object.keys(_Axis).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _Axis[key];
});

var _Channel = require("./types/Channel");

Object.keys(_Channel).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _Channel[key];
});

var _ChannelDef2 = require("./types/ChannelDef");

Object.keys(_ChannelDef2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _ChannelDef2[key];
});

var _Data = require("./types/Data");

Object.keys(_Data).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _Data[key];
});

var _Encoding = require("./types/Encoding");

Object.keys(_Encoding).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _Encoding[key];
});

var _Legend = require("./types/Legend");

Object.keys(_Legend).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _Legend[key];
});

var _Scale = require("./types/Scale");

Object.keys(_Scale).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _Scale[key];
});

var _VegaLite = require("./types/VegaLite");

Object.keys(_VegaLite).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  exports[key] = _VegaLite[key];
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
