// @superset-ui/legacy-plugin-chart-calendar/lib/Calendar.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _d3Array = require("d3-array");

var _d3Selection = require("d3-selection");

var _color = require("@superset-ui/color");

var _numberFormat = require("@superset-ui/number-format");

var _timeFormat = require("@superset-ui/time-format");

var _calHeatmap = _interopRequireDefault(require("./vendor/cal-heatmap"));

require("./vendor/cal-heatmap.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
function convertUTC(dttm) {
  return new Date(dttm.getUTCFullYear(), dttm.getUTCMonth(), dttm.getUTCDate(), dttm.getUTCHours(), dttm.getUTCMinutes(), dttm.getUTCSeconds());
}

const convertUTCTS = uts => convertUTC(new Date(uts)).getTime();

const propTypes = {
  data: _propTypes.default.shape({
    // Object hashed by metric name,
    // then hashed by timestamp (in seconds, not milliseconds) as float
    // the innermost value is count
    // e.g. { count_distinct_something: { 1535034236.0: 3 } }
    data: _propTypes.default.object,
    domain: _propTypes.default.string,
    range: _propTypes.default.number,
    // timestamp in milliseconds
    start: _propTypes.default.number,
    subdomain: _propTypes.default.string
  }),
  height: _propTypes.default.number,
  // eslint-disable-next-line react/sort-prop-types
  cellPadding: _propTypes.default.number,
  // eslint-disable-next-line react/sort-prop-types
  cellRadius: _propTypes.default.number,
  // eslint-disable-next-line react/sort-prop-types
  cellSize: _propTypes.default.number,
  linearColorScheme: _propTypes.default.string,
  showLegend: _propTypes.default.bool,
  showMetricName: _propTypes.default.bool,
  showValues: _propTypes.default.bool,
  steps: _propTypes.default.number,
  timeFormat: _propTypes.default.string,
  valueFormat: _propTypes.default.string,
  verboseMap: _propTypes.default.object
};

function Calendar(element, props) {
  const {
    data,
    height,
    cellPadding = 3,
    cellRadius = 0,
    cellSize = 10,
    domainGranularity,
    linearColorScheme,
    showLegend,
    showMetricName,
    showValues,
    steps,
    subdomainGranularity,
    timeFormat,
    valueFormat,
    verboseMap
  } = props;
  const valueFormatter = (0, _numberFormat.getNumberFormatter)(valueFormat);
  const timeFormatter = (0, _timeFormat.getTimeFormatter)(timeFormat);
  const container = (0, _d3Selection.select)(element).classed('superset-legacy-chart-calendar', true).style('height', height);
  container.selectAll('*').remove();
  const div = container.append('div');
  const subDomainTextFormat = showValues ? (date, value) => valueFormatter(value) : null; // Trick to convert all timestamps to UTC
  // TODO: Verify if this conversion is really necessary
  // since all timestamps should always be in UTC.

  const metricsData = {};
  Object.keys(data.data).forEach(metric => {
    metricsData[metric] = {};
    Object.keys(data.data[metric]).forEach(ts => {
      metricsData[metric][convertUTCTS(ts * 1000) / 1000] = data.data[metric][ts];
    });
  });
  Object.keys(metricsData).forEach(metric => {
    const calContainer = div.append('div');

    if (showMetricName) {
      calContainer.text("Metric: " + (verboseMap[metric] || metric));
    }

    const timestamps = metricsData[metric];
    const extents = (0, _d3Array.extent)(Object.keys(timestamps), key => timestamps[key]);
    const step = (extents[1] - extents[0]) / (steps - 1);
    const colorScale = (0, _color.getSequentialSchemeRegistry)().get(linearColorScheme).createLinearScale(extents);
    const legend = (0, _d3Array.range)(steps).map(i => extents[0] + step * i);
    const legendColors = legend.map(x => colorScale(x));
    const cal = new _calHeatmap.default();
    cal.init({
      start: convertUTCTS(data.start),
      data: timestamps,
      itemSelector: calContainer.node(),
      legendVerticalPosition: 'top',
      cellSize,
      cellPadding,
      cellRadius,
      legendCellSize: cellSize,
      legendCellPadding: 2,
      legendCellRadius: cellRadius,
      tooltip: true,
      domain: domainGranularity,
      subDomain: subdomainGranularity,
      range: data.range,
      browsing: true,
      legend,
      legendColors: {
        colorScale,
        min: legendColors[0],
        max: legendColors[legendColors.length - 1],
        empty: 'white'
      },
      displayLegend: showLegend,
      itemName: '',
      valueFormatter,
      timeFormatter,
      subDomainTextFormat
    });
  });
}

Calendar.displayName = 'Calendar';
Calendar.propTypes = propTypes;
var _default = Calendar;
exports.default = _default;
