// encodable/lib/typeGuards/Scale.js
"use strict";

exports.__esModule = true;
exports.isContinuousScaleConfig = isContinuousScaleConfig;
exports.isDiscretizingScaleConfig = isDiscretizingScaleConfig;
exports.isScaleConfigWithZero = isScaleConfigWithZero;
exports.isCategoricalColorScale = isCategoricalColorScale;
exports.isD3Scale = isD3Scale;
exports.isContinuousScale = isContinuousScale;
exports.isDiscretizingScale = isDiscretizingScale;
exports.isTimeScale = isTimeScale;
exports.isSchemeParams = isSchemeParams;

var _color = require("@superset-ui/color");

var _scaleCategories = require("../parsers/scale/scaleCategories");

var _isPropertySupportedByScaleType = _interopRequireDefault(require("../parsers/scale/isPropertySupportedByScaleType"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function isContinuousScaleConfig(config) {
  return _scaleCategories.continuousScaleTypesSet.has(config.type);
}

function isDiscretizingScaleConfig(config) {
  return _scaleCategories.discretizingScaleTypesSet.has(config.type);
}

function isScaleConfigWithZero(config) {
  return (0, _isPropertySupportedByScaleType.default)('zero', config.type);
}

function isCategoricalColorScale(scale) {
  return scale instanceof _color.CategoricalColorScale;
}

function isD3Scale(scale) {
  return !isCategoricalColorScale(scale);
}

function isContinuousScale(scale, scaleType) {
  return scale && _scaleCategories.continuousScaleTypesSet.has(scaleType);
}

function isDiscretizingScale(scale, scaleType) {
  return scale && _scaleCategories.discretizingScaleTypesSet.has(scaleType);
}

function isTimeScale(scale, scaleType) {
  return scale && _scaleCategories.timeScaleTypesSet.has(scaleType);
}

function isSchemeParams(scheme) {
  return Object.prototype.toString.call(scheme) !== '[object String]' && !!scheme.name;
}
