// @vx/shape/esm/shapes/Pie.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/* eslint-disable @typescript-eslint/unbound-method */
import React from 'react';
import cx from 'classnames';
import { Group } from '@vx/group';
import { arc as d3Arc, pie as d3Pie } from 'd3-shape';
import setNumOrAccessor from '../util/setNumberOrNumberAccessor';
export default function Pie(_ref) {
  var className = _ref.className,
      top = _ref.top,
      left = _ref.left,
      _ref$data = _ref.data,
      data = _ref$data === void 0 ? [] : _ref$data,
      centroid = _ref.centroid,
      _ref$innerRadius = _ref.innerRadius,
      innerRadius = _ref$innerRadius === void 0 ? 0 : _ref$innerRadius,
      outerRadius = _ref.outerRadius,
      cornerRadius = _ref.cornerRadius,
      startAngle = _ref.startAngle,
      endAngle = _ref.endAngle,
      padAngle = _ref.padAngle,
      padRadius = _ref.padRadius,
      pieSort = _ref.pieSort,
      pieSortValues = _ref.pieSortValues,
      pieValue = _ref.pieValue,
      children = _ref.children,
      restProps = _objectWithoutPropertiesLoose(_ref, ["className", "top", "left", "data", "centroid", "innerRadius", "outerRadius", "cornerRadius", "startAngle", "endAngle", "padAngle", "padRadius", "pieSort", "pieSortValues", "pieValue", "children"]);

  var path = d3Arc();
  if (innerRadius != null) setNumOrAccessor(path.innerRadius, innerRadius);
  if (outerRadius != null) setNumOrAccessor(path.outerRadius, outerRadius);
  if (cornerRadius != null) setNumOrAccessor(path.cornerRadius, cornerRadius);
  if (padRadius != null) setNumOrAccessor(path.padRadius, padRadius);
  var pie = d3Pie(); // ts can't distinguish between these method overloads

  if (pieSort === null) pie.sort(pieSort);else if (pieSort != null) pie.sort(pieSort);
  if (pieSortValues === null) pie.sortValues(pieSortValues);else if (pieSortValues != null) pie.sortValues(pieSortValues);
  if (pieValue != null) pie.value(pieValue);
  if (padAngle != null) setNumOrAccessor(pie.padAngle, padAngle);
  if (startAngle != null) setNumOrAccessor(pie.startAngle, startAngle);
  if (endAngle != null) setNumOrAccessor(pie.endAngle, endAngle);
  var arcs = pie(data);
  if (children) return /*#__PURE__*/React.createElement(React.Fragment, null, children({
    arcs: arcs,
    path: path,
    pie: pie
  }));
  return /*#__PURE__*/React.createElement(Group, {
    className: "vx-pie-arcs-group",
    top: top,
    left: left
  }, arcs.map(function (arc, i) {
    return /*#__PURE__*/React.createElement("g", {
      key: "pie-arc-" + i
    }, /*#__PURE__*/React.createElement("path", _extends({
      className: cx('vx-pie-arc', className),
      d: path(arc) || ''
    }, restProps)), centroid && centroid(path.centroid(arc), arc));
  }));
}
Pie.propTypes = {
  className: _pt.string,
  top: _pt.number,
  left: _pt.number,
  data: _pt.array,
  pieSort: _pt.oneOfType([_pt.oneOf([null]), _pt.func]),
  pieSortValues: _pt.oneOfType([_pt.oneOf([null]), _pt.func]),
  centroid: _pt.func,
  children: _pt.func
};
