// viewport-mercator-project/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "default", {
  enumerable: true,
  get: function get() {
    return _webMercatorViewport.default;
  }
});
Object.defineProperty(exports, "WebMercatorViewport", {
  enumerable: true,
  get: function get() {
    return _webMercatorViewport.default;
  }
});
Object.defineProperty(exports, "PerspectiveMercatorViewport", {
  enumerable: true,
  get: function get() {
    return _webMercatorViewport.default;
  }
});
Object.defineProperty(exports, "fitBounds", {
  enumerable: true,
  get: function get() {
    return _fitBounds.default;
  }
});
Object.defineProperty(exports, "normalizeViewportProps", {
  enumerable: true,
  get: function get() {
    return _normalizeViewportProps.default;
  }
});
Object.defineProperty(exports, "flyToViewport", {
  enumerable: true,
  get: function get() {
    return _flyToViewport.default;
  }
});
Object.defineProperty(exports, "lngLatToWorld", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.lngLatToWorld;
  }
});
Object.defineProperty(exports, "worldToLngLat", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.worldToLngLat;
  }
});
Object.defineProperty(exports, "worldToPixels", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.worldToPixels;
  }
});
Object.defineProperty(exports, "pixelsToWorld", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.pixelsToWorld;
  }
});
Object.defineProperty(exports, "getMeterZoom", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getMeterZoom;
  }
});
Object.defineProperty(exports, "getDistanceScales", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getDistanceScales;
  }
});
Object.defineProperty(exports, "addMetersToLngLat", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.addMetersToLngLat;
  }
});
Object.defineProperty(exports, "getViewMatrix", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getViewMatrix;
  }
});
Object.defineProperty(exports, "getProjectionMatrix", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getProjectionMatrix;
  }
});
Object.defineProperty(exports, "getProjectionParameters", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getProjectionParameters;
  }
});
Object.defineProperty(exports, "getUncenteredViewMatrix", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.getViewMatrix;
  }
});
Object.defineProperty(exports, "projectFlat", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.lngLatToWorld;
  }
});
Object.defineProperty(exports, "unprojectFlat", {
  enumerable: true,
  get: function get() {
    return _webMercatorUtils.worldToLngLat;
  }
});

var _webMercatorViewport = _interopRequireDefault(require("./web-mercator-viewport"));

var _fitBounds = _interopRequireDefault(require("./fit-bounds"));

var _normalizeViewportProps = _interopRequireDefault(require("./normalize-viewport-props"));

var _flyToViewport = _interopRequireDefault(require("./fly-to-viewport"));

var _webMercatorUtils = require("./web-mercator-utils");
