// @superset-ui/time-format/lib/TimeRangeFormatter.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _core = require("@superset-ui/core");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class TimeRangeFormatter extends _core.ExtensibleFunction {
  constructor(config) {
    super(value => this.format(value));

    _defineProperty(this, "id", void 0);

    _defineProperty(this, "label", void 0);

    _defineProperty(this, "description", void 0);

    _defineProperty(this, "formatFunc", void 0);

    _defineProperty(this, "useLocalTime", void 0);

    const {
      id,
      label,
      description = '',
      formatFunc,
      useLocalTime = false
    } = config;
    this.id = id;
    this.label = label != null ? label : id;
    this.description = description;
    this.formatFunc = formatFunc;
    this.useLocalTime = useLocalTime;
  }

  format(values) {
    return this.formatFunc(values);
  }

}

var _default = TimeRangeFormatter;
exports.default = _default;
