// @superset-ui/time-format/lib/TimeFormatterRegistrySingleton.js
"use strict";

exports.__esModule = true;
exports.getTimeRangeFormatter = getTimeRangeFormatter;
exports.formatTimeRange = formatTimeRange;
exports.getTimeFormatter = getTimeFormatter;
exports.getTimeFormatterForGranularity = getTimeFormatterForGranularity;
exports.formatTime = formatTime;
exports.default = void 0;

var _core = require("@superset-ui/core");

var _TimeFormatterRegistry = _interopRequireDefault(require("./TimeFormatterRegistry"));

var _TimeFormatter = _interopRequireDefault(require("./TimeFormatter"));

var _TimeFormatsForGranularity = _interopRequireDefault(require("./TimeFormatsForGranularity"));

var _TimeFormats = require("./TimeFormats");

var _createTimeRangeFromGranularity = _interopRequireDefault(require("./utils/createTimeRangeFromGranularity"));

var _TimeRangeFormatter = _interopRequireDefault(require("./TimeRangeFormatter"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const getInstance = (0, _core.makeSingleton)(_TimeFormatterRegistry.default);
var _default = getInstance;
exports.default = _default;

function getTimeRangeFormatter(formatId) {
  return new _TimeRangeFormatter.default({
    id: formatId || 'undefined',
    formatFunc: range => {
      const format = getInstance().get(formatId);
      const [start, end] = range.map(value => format(value));
      return start === end ? start : [start, end].join(' — ');
    },
    useLocalTime: formatId == null ? void 0 : formatId.startsWith(_TimeFormats.LOCAL_PREFIX)
  });
}

function formatTimeRange(formatId, range) {
  return getTimeRangeFormatter(formatId)(range);
}

function getTimeFormatter(formatId, granularity) {
  if (granularity) {
    const formatString = formatId || _TimeFormatsForGranularity.default[granularity];
    const timeRangeFormatter = getTimeRangeFormatter(formatString);
    return new _TimeFormatter.default({
      id: [formatString, granularity].join('/'),
      formatFunc: value => timeRangeFormatter.format((0, _createTimeRangeFromGranularity.default)(value, granularity, timeRangeFormatter.useLocalTime)),
      useLocalTime: timeRangeFormatter.useLocalTime
    });
  }

  return getInstance().get(formatId);
}
/**
 * Syntactic sugar for backward compatibility
 * TODO: Deprecate this in the next breaking change.
 * @param granularity
 */


function getTimeFormatterForGranularity(granularity) {
  return getTimeFormatter(undefined, granularity);
}

function formatTime(formatId, value, granularity) {
  return getTimeFormatter(formatId, granularity)(value);
}
