// @data-ui/event-flow/src/constants.js
// event attributes
export const TS0 = 'TS0';
export const TS_NEXT = 'TS_NEXT';
export const TS_PREV = 'TS_PREV';
export const TS = 'TS';

export const EVENT_NAME = 'EVENT_NAME';
export const ENTITY_ID = 'ENTITY_ID';
export const EVENT_UUID = 'EVENT_UUID';

export const META = 'META';
export const ELAPSED_MS_ROOT = 'ELAPSED_MS_ROOT';
export const ELAPSED_MS = 'ELAPSED_MS';
export const EVENT_COUNT = 'EVENT_COUNT';

export const ANY_EVENT_TYPE = 'ANY_EVENT_TYPE';

// scales
export const ELAPSED_TIME_SCALE = 'ELAPSED_TIME_SCALE';
export const EVENT_SEQUENCE_SCALE = 'EVENT_SEQUENCE_SCALE';
export const NODE_COLOR_SCALE = 'NODE_COLOR_SCALE';
export const EVENT_COUNT_SCALE = 'EVENT_COUNT_SCALE';
export const NODE_SEQUENCE_SCALE = 'NODE_SEQUENCE_SCALE';

// node sorters
export const ORDER_BY_EVENT_COUNT = 'ORDER_BY_EVENT_COUNT';
export const ORDER_BY_ELAPSED_MS = 'ORDER_BY_ELAPSED_MS';

// note this can't have spaces or it breaks its use in the Pattern 'url(#id)''
export const FILTERED_EVENTS = 'FILTERED_EVENTS';
export const CLIP_ID = 'CLIP_PATH_ID';
