// math.gl/dist/es5/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Vector2", {
  enumerable: true,
  get: function get() {
    return _vector.default;
  }
});
Object.defineProperty(exports, "Vector3", {
  enumerable: true,
  get: function get() {
    return _vector2.default;
  }
});
Object.defineProperty(exports, "Vector4", {
  enumerable: true,
  get: function get() {
    return _vector3.default;
  }
});
Object.defineProperty(exports, "Matrix4", {
  enumerable: true,
  get: function get() {
    return _matrix.default;
  }
});
Object.defineProperty(exports, "Quaternion", {
  enumerable: true,
  get: function get() {
    return _quaternion.default;
  }
});
Object.defineProperty(exports, "config", {
  enumerable: true,
  get: function get() {
    return _common.config;
  }
});
Object.defineProperty(exports, "checkNumber", {
  enumerable: true,
  get: function get() {
    return _common.checkNumber;
  }
});
Object.defineProperty(exports, "configure", {
  enumerable: true,
  get: function get() {
    return _common.configure;
  }
});
Object.defineProperty(exports, "formatValue", {
  enumerable: true,
  get: function get() {
    return _common.formatValue;
  }
});
Object.defineProperty(exports, "isArray", {
  enumerable: true,
  get: function get() {
    return _common.isArray;
  }
});
Object.defineProperty(exports, "clone", {
  enumerable: true,
  get: function get() {
    return _common.clone;
  }
});
Object.defineProperty(exports, "radians", {
  enumerable: true,
  get: function get() {
    return _common.radians;
  }
});
Object.defineProperty(exports, "degrees", {
  enumerable: true,
  get: function get() {
    return _common.degrees;
  }
});
Object.defineProperty(exports, "sin", {
  enumerable: true,
  get: function get() {
    return _common.sin;
  }
});
Object.defineProperty(exports, "cos", {
  enumerable: true,
  get: function get() {
    return _common.cos;
  }
});
Object.defineProperty(exports, "tan", {
  enumerable: true,
  get: function get() {
    return _common.tan;
  }
});
Object.defineProperty(exports, "asin", {
  enumerable: true,
  get: function get() {
    return _common.asin;
  }
});
Object.defineProperty(exports, "acos", {
  enumerable: true,
  get: function get() {
    return _common.acos;
  }
});
Object.defineProperty(exports, "atan", {
  enumerable: true,
  get: function get() {
    return _common.atan;
  }
});
Object.defineProperty(exports, "clamp", {
  enumerable: true,
  get: function get() {
    return _common.clamp;
  }
});
Object.defineProperty(exports, "lerp", {
  enumerable: true,
  get: function get() {
    return _common.lerp;
  }
});
Object.defineProperty(exports, "equals", {
  enumerable: true,
  get: function get() {
    return _common.equals;
  }
});
Object.defineProperty(exports, "_SphericalCoordinates", {
  enumerable: true,
  get: function get() {
    return _sphericalCoordinates.default;
  }
});
Object.defineProperty(exports, "_Pose", {
  enumerable: true,
  get: function get() {
    return _pose.default;
  }
});
Object.defineProperty(exports, "_Euler", {
  enumerable: true,
  get: function get() {
    return _euler.default;
  }
});
Object.defineProperty(exports, "_Polygon", {
  enumerable: true,
  get: function get() {
    return _polygon.default;
  }
});

var _vector = _interopRequireDefault(require("./vector2"));

var _vector2 = _interopRequireDefault(require("./vector3"));

var _vector3 = _interopRequireDefault(require("./vector4"));

var _matrix = _interopRequireDefault(require("./matrix4"));

var _quaternion = _interopRequireDefault(require("./quaternion"));

var _common = require("./lib/common");

var _sphericalCoordinates = _interopRequireDefault(require("./spherical-coordinates"));

var _pose = _interopRequireDefault(require("./pose"));

var _euler = _interopRequireDefault(require("./euler"));

var _polygon = _interopRequireDefault(require("./addons/polygon"));
