// @superset-ui/legacy-plugin-chart-pivot-table/lib/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _translation = require("@superset-ui/translation");

var _chartControls = require("@superset-ui/chart-controls");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
var _default = {
  controlPanelSections: [{
    label: (0, _translation.t)('Query'),
    expanded: true,
    controlSetRows: [['metrics'], ['adhoc_filters'], ['groupby'], ['columns'], ['row_limit', null]]
  }, {
    label: (0, _translation.t)('Pivot Options'),
    controlSetRows: [[{
      name: 'pandas_aggfunc',
      config: {
        type: 'SelectControl',
        label: (0, _translation.t)('Aggregation function'),
        clearable: false,
        choices: (0, _chartControls.formatSelectOptions)(['sum', 'mean', 'min', 'max', 'std', 'var']),
        default: 'sum',
        description: (0, _translation.t)('Aggregate function to apply when pivoting and ' + 'computing the total rows and columns')
      }
    }, {
      name: 'pivot_margins',
      config: {
        type: 'CheckboxControl',
        label: (0, _translation.t)('Show totals'),
        renderTrigger: false,
        default: true,
        description: (0, _translation.t)('Display total row/column')
      }
    }], [{
      name: 'number_format',
      config: {
        type: 'SelectControl',
        freeForm: true,
        label: (0, _translation.t)('Number format'),
        renderTrigger: true,
        default: 'SMART_NUMBER',
        choices: _chartControls.D3_FORMAT_OPTIONS,
        description: _chartControls.D3_FORMAT_DOCS
      }
    }, {
      name: 'combine_metric',
      config: {
        type: 'CheckboxControl',
        label: (0, _translation.t)('Combine Metrics'),
        default: false,
        description: (0, _translation.t)('Display metrics side by side within each column, as ' + 'opposed to each column being displayed side by side for each metric.')
      }
    }], [{
      name: 'transpose_pivot',
      config: {
        type: 'CheckboxControl',
        label: (0, _translation.t)('Transpose Pivot'),
        default: false,
        description: (0, _translation.t)('Swap Groups and Columns')
      }
    }]]
  }],
  controlOverrides: {
    groupby: {
      includeTime: true
    },
    columns: {
      includeTime: true
    }
  },
  sectionOverrides: {
    druidTimeSeries: {
      controlSetRows: [['granularity', 'druid_time_origin'], ['time_range']]
    },
    sqlaTimeSeries: {
      controlSetRows: [['granularity_sqla', 'time_grain_sqla'], ['time_range']]
    }
  }
};
exports.default = _default;
