// @superset-ui/plugin-chart-word-cloud/lib/chart/WordCloud.js
"use strict";

exports.__esModule = true;
exports.default = exports.ROTATION = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _d3Cloud = _interopRequireDefault(require("d3-cloud"));

var _encodable = require("encodable");

var _style = require("@superset-ui/style");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const ROTATION = {
  flat: () => 0,
  // this calculates a random rotation between -90 and 90 degrees.
  random: () => Math.floor(Math.random() * 6 - 3) * 30,
  square: () => Math.floor(Math.random() * 2) * 90
};
exports.ROTATION = ROTATION;
const defaultProps = {
  encoding: {},
  rotation: 'flat'
};

class WordCloud extends _react.default.PureComponent {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "isComponentMounted", false);

    _defineProperty(this, "state", {
      words: []
    });

    _defineProperty(this, "wordCloudEncoderFactory", (0, _encodable.createEncoderFactory)({
      channelTypes: {
        color: 'Color',
        fontFamily: 'Category',
        fontSize: 'Numeric',
        fontWeight: 'Category',
        text: 'Text'
      },
      defaultEncoding: {
        color: {
          value: 'black'
        },
        fontFamily: {
          value: this.props.theme.typography.families.sansSerif
        },
        fontSize: {
          value: 20
        },
        fontWeight: {
          value: 'bold'
        },
        text: {
          value: ''
        }
      }
    }));

    _defineProperty(this, "createEncoder", this.wordCloudEncoderFactory.createSelector());

    _defineProperty(this, "setWords", words => {
      if (this.isComponentMounted) {
        this.setState({
          words
        });
      }
    });
  }

  componentDidMount() {
    this.isComponentMounted = true;
    this.update();
  }

  componentDidUpdate(prevProps) {
    const {
      data,
      encoding,
      width,
      height,
      rotation
    } = this.props;

    if (prevProps.data !== data || prevProps.encoding !== encoding || prevProps.width !== width || prevProps.height !== height || prevProps.rotation !== rotation) {
      this.update();
    }
  }

  componentWillUnmount() {
    this.isComponentMounted = false;
  }

  update() {
    const {
      data,
      width,
      height,
      rotation,
      encoding
    } = this.props;
    const encoder = this.createEncoder(encoding);
    encoder.setDomainFromDataset(data);
    (0, _d3Cloud.default)().size([width, height]) // clone the data because cloudLayout mutates input
    .words(data.map(d => _extends({}, d))).padding(5).rotate(ROTATION[rotation] || ROTATION.flat).text(d => encoder.channels.text.getValueFromDatum(d)).font(d => encoder.channels.fontFamily.encodeDatum(d, this.props.theme.typography.families.sansSerif)).fontWeight(d => encoder.channels.fontWeight.encodeDatum(d, 'normal')).fontSize(d => encoder.channels.fontSize.encodeDatum(d, 0)).on('end', this.setWords).start();
  }

  render() {
    const {
      width,
      height,
      encoding
    } = this.props;
    const {
      words
    } = this.state;
    const encoder = this.createEncoder(encoding);
    encoder.channels.color.setDomainFromDataset(words);
    return /*#__PURE__*/_react.default.createElement("svg", {
      width: width,
      height: height
    }, /*#__PURE__*/_react.default.createElement("g", {
      transform: "translate(" + width / 2 + "," + height / 2 + ")"
    }, words.map(w => /*#__PURE__*/_react.default.createElement("text", {
      key: w.text,
      fontSize: w.size + "px",
      fontWeight: w.weight,
      fontFamily: w.font,
      fill: encoder.channels.color.encodeDatum(w, ''),
      textAnchor: "middle",
      transform: "translate(" + w.x + ", " + w.y + ") rotate(" + w.rotate + ")"
    }, w.text))));
  }

}

_defineProperty(WordCloud, "propTypes", {
  rotation: _propTypes.default.any,
  data: _propTypes.default.array.isRequired,
  height: _propTypes.default.number.isRequired,
  width: _propTypes.default.number.isRequired
});

_defineProperty(WordCloud, "defaultProps", defaultProps);

var _default = (0, _style.withTheme)(WordCloud);

exports.default = _default;
