// @data-ui/xy-chart/lib/glyph/GlyphDotComponent.js
"use strict";

exports.__esModule = true;
exports.default = GlyphDotComponent;

var _react = _interopRequireDefault(require("react"));

var _glyph = require("@vx/glyph");

var _propShapes = require("../utils/propShapes");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var defaultPropTypes = {
  onMouseMove: null,
  onMouseLeave: null,
  strokeDasharray: null
};

function GlyphDotComponent(_ref) {
  var x = _ref.x,
      y = _ref.y,
      size = _ref.size,
      fill = _ref.fill,
      fillOpacity = _ref.fillOpacity,
      stroke = _ref.stroke,
      strokeWidth = _ref.strokeWidth,
      strokeDasharray = _ref.strokeDasharray,
      onClick = _ref.onClick,
      onMouseMove = _ref.onMouseMove,
      onMouseLeave = _ref.onMouseLeave,
      data = _ref.data,
      datum = _ref.datum;
  return _react.default.createElement(_glyph.GlyphDot, {
    cx: x,
    cy: y,
    r: size,
    fill: fill,
    fillOpacity: fillOpacity,
    stroke: stroke,
    strokeWidth: strokeWidth,
    strokeDasharray: strokeDasharray,
    onClick: onClick && function (event) {
      onClick({
        event: event,
        data: data,
        datum: datum,
        color: fill
      });
    },
    onMouseMove: onMouseMove && function (event) {
      onMouseMove({
        event: event,
        data: data,
        datum: datum,
        color: fill
      });
    },
    onMouseLeave: onMouseLeave
  });
}

GlyphDotComponent.propTypes = _extends({}, _propShapes.pointComponentPropTypes);
GlyphDotComponent.defaultProps = defaultPropTypes;
