// @superset-ui/legacy-plugin-chart-parallel-coordinates/esm/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _translation = require("@superset-ui/translation");

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
var _default = {
  controlPanelSections: [{
    label: (0, _translation.t)('Query'),
    expanded: true,
    controlSetRows: [['series'], ['metrics'], ['secondary_metric'], ['adhoc_filters'], ['limit', 'row_limit']]
  }, {
    label: (0, _translation.t)('Options'),
    expanded: true,
    controlSetRows: [[{
      name: 'show_datatable',
      config: {
        type: 'CheckboxControl',
        label: (0, _translation.t)('Data Table'),
        default: false,
        renderTrigger: true,
        description: (0, _translation.t)('Whether to display the interactive data table')
      }
    }, {
      name: 'include_series',
      config: {
        type: 'CheckboxControl',
        label: (0, _translation.t)('Include Series'),
        renderTrigger: true,
        default: false,
        description: (0, _translation.t)('Include series name as an axis')
      }
    }], ['linear_color_scheme']]
  }]
};
exports.default = _default;
