// math.gl/dist/gl-matrix-min/quat.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _fromMat = require('gl-quat/fromMat3');

Object.defineProperty(exports, 'quat_fromMat3', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_fromMat).default;
  }
});

var _identity = require('gl-quat/identity');

Object.defineProperty(exports, 'quat_identity', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_identity).default;
  }
});

var _length = require('gl-quat/length');

Object.defineProperty(exports, 'quat_length', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_length).default;
  }
});

var _squaredLength = require('gl-quat/squaredLength');

Object.defineProperty(exports, 'quat_squaredLength', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_squaredLength).default;
  }
});

var _dot = require('gl-quat/dot');

Object.defineProperty(exports, 'quat_dot', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dot).default;
  }
});

var _rotationTo = require('gl-quat/rotationTo');

Object.defineProperty(exports, 'quat_rotationTo', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_rotationTo).default;
  }
});

var _add = require('gl-quat/add');

Object.defineProperty(exports, 'quat_add', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_add).default;
  }
});

var _calculateW = require('gl-quat/calculateW');

Object.defineProperty(exports, 'quat_calculateW', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_calculateW).default;
  }
});

var _conjugate = require('gl-quat/conjugate');

Object.defineProperty(exports, 'quat_conjugate', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_conjugate).default;
  }
});

var _invert = require('gl-quat/invert');

Object.defineProperty(exports, 'quat_invert', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_invert).default;
  }
});

var _lerp = require('gl-quat/lerp');

Object.defineProperty(exports, 'quat_lerp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_lerp).default;
  }
});

var _multiply = require('gl-quat/multiply');

Object.defineProperty(exports, 'quat_multiply', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_multiply).default;
  }
});

var _normalize = require('gl-quat/normalize');

Object.defineProperty(exports, 'quat_normalize', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_normalize).default;
  }
});

var _rotateX = require('gl-quat/rotateX');

Object.defineProperty(exports, 'quat_rotateX', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_rotateX).default;
  }
});

var _rotateY = require('gl-quat/rotateY');

Object.defineProperty(exports, 'quat_rotateY', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_rotateY).default;
  }
});

var _rotateZ = require('gl-quat/rotateZ');

Object.defineProperty(exports, 'quat_rotateZ', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_rotateZ).default;
  }
});

var _scale = require('gl-quat/scale');

Object.defineProperty(exports, 'quat_scale', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scale).default;
  }
});

var _set = require('gl-quat/set');

Object.defineProperty(exports, 'quat_set', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_set).default;
  }
});

var _setAxisAngle = require('gl-quat/setAxisAngle');

Object.defineProperty(exports, 'quat_setAxisAngle', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_setAxisAngle).default;
  }
});

var _slerp = require('gl-quat/slerp');

Object.defineProperty(exports, 'quat_slerp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_slerp).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
