// math.gl/dist/gl-matrix-min/vec4.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _set = require('gl-vec4/set');

Object.defineProperty(exports, 'vec4_set', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_set).default;
  }
});

var _distance = require('gl-vec4/distance');

Object.defineProperty(exports, 'vec4_distance', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_distance).default;
  }
});

var _add = require('gl-vec4/add');

Object.defineProperty(exports, 'vec4_add', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_add).default;
  }
});

var _subtract = require('gl-vec4/subtract');

Object.defineProperty(exports, 'vec4_subtract', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_subtract).default;
  }
});

var _multiply = require('gl-vec4/multiply');

Object.defineProperty(exports, 'vec4_multiply', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_multiply).default;
  }
});

var _divide = require('gl-vec4/divide');

Object.defineProperty(exports, 'vec4_divide', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_divide).default;
  }
});

var _scale = require('gl-vec4/scale');

Object.defineProperty(exports, 'vec4_scale', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scale).default;
  }
});

var _scaleAndAdd = require('gl-vec4/scaleAndAdd');

Object.defineProperty(exports, 'vec4_scaleAndAdd', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scaleAndAdd).default;
  }
});

var _negate = require('gl-vec4/negate');

Object.defineProperty(exports, 'vec4_negate', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_negate).default;
  }
});

var _inverse = require('gl-vec4/inverse');

Object.defineProperty(exports, 'vec4_inverse', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_inverse).default;
  }
});

var _normalize = require('gl-vec4/normalize');

Object.defineProperty(exports, 'vec4_normalize', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_normalize).default;
  }
});

var _dot = require('gl-vec4/dot');

Object.defineProperty(exports, 'vec4_dot', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dot).default;
  }
});

var _lerp = require('gl-vec4/lerp');

Object.defineProperty(exports, 'vec4_lerp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_lerp).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
