// @vx/responsive/lib/enhancers/withScreenSize.js
"use strict";

exports.__esModule = true;
exports.default = withScreenSize;

var _debounce = _interopRequireDefault(require("lodash/debounce"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function withScreenSize(BaseComponent) {
  var WrappedComponent =
  /*#__PURE__*/
  function (_React$Component) {
    _inheritsLoose(WrappedComponent, _React$Component);

    function WrappedComponent(props) {
      var _this;

      _this = _React$Component.call(this, props) || this;
      _this.state = {
        screenWidth: null,
        screenHeight: null
      };
      _this.handleResize = (0, _debounce.default)(_this.resize.bind(_assertThisInitialized(_this)), props.windowResizeDebounceTime).bind(_assertThisInitialized(_this));
      return _this;
    }

    var _proto = WrappedComponent.prototype;

    _proto.componentDidMount = function componentDidMount() {
      window.addEventListener('resize', this.handleResize, false);
      this.resize();
    };

    _proto.componentWillUnmount = function componentWillUnmount() {
      window.removeEventListener('resize', this.handleResize, false);
    };

    _proto.resize = function resize()
    /** event */
    {
      this.setState(function ()
      /** prevState, props */
      {
        return {
          screenWidth: window.innerWidth,
          screenHeight: window.innerHeight
        };
      });
    };

    _proto.render = function render() {
      var _this$state = this.state,
          screenWidth = _this$state.screenWidth,
          screenHeight = _this$state.screenHeight;
      if (!screenWidth && !screenHeight) return null;
      return _react.default.createElement(BaseComponent, _extends({
        screenWidth: screenWidth,
        screenHeight: screenHeight
      }, this.props));
    };

    return WrappedComponent;
  }(_react.default.Component);

  WrappedComponent.propTypes = {
    windowResizeDebounceTime: _propTypes.default.number
  };
  WrappedComponent.defaultProps = {
    windowResizeDebounceTime: 300
  };
  return WrappedComponent;
}
