// react-syntax-highlighter/dist/languages/hljs/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _c = require('./1c');

Object.defineProperty(exports, 'oneC', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_c).default;
  }
});

var _abnf = require('./abnf');

Object.defineProperty(exports, 'abnf', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_abnf).default;
  }
});

var _accesslog = require('./accesslog');

Object.defineProperty(exports, 'accesslog', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_accesslog).default;
  }
});

var _actionscript = require('./actionscript');

Object.defineProperty(exports, 'actionscript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_actionscript).default;
  }
});

var _ada = require('./ada');

Object.defineProperty(exports, 'ada', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ada).default;
  }
});

var _apache = require('./apache');

Object.defineProperty(exports, 'apache', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_apache).default;
  }
});

var _applescript = require('./applescript');

Object.defineProperty(exports, 'applescript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_applescript).default;
  }
});

var _arduino = require('./arduino');

Object.defineProperty(exports, 'arduino', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_arduino).default;
  }
});

var _armasm = require('./armasm');

Object.defineProperty(exports, 'armasm', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_armasm).default;
  }
});

var _asciidoc = require('./asciidoc');

Object.defineProperty(exports, 'asciidoc', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_asciidoc).default;
  }
});

var _aspectj = require('./aspectj');

Object.defineProperty(exports, 'aspectj', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_aspectj).default;
  }
});

var _autohotkey = require('./autohotkey');

Object.defineProperty(exports, 'autohotkey', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_autohotkey).default;
  }
});

var _autoit = require('./autoit');

Object.defineProperty(exports, 'autoit', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_autoit).default;
  }
});

var _avrasm = require('./avrasm');

Object.defineProperty(exports, 'avrasm', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_avrasm).default;
  }
});

var _awk = require('./awk');

Object.defineProperty(exports, 'awk', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_awk).default;
  }
});

var _axapta = require('./axapta');

Object.defineProperty(exports, 'axapta', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_axapta).default;
  }
});

var _bash = require('./bash');

Object.defineProperty(exports, 'bash', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_bash).default;
  }
});

var _basic = require('./basic');

Object.defineProperty(exports, 'basic', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_basic).default;
  }
});

var _bnf = require('./bnf');

Object.defineProperty(exports, 'bnf', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_bnf).default;
  }
});

var _brainfuck = require('./brainfuck');

Object.defineProperty(exports, 'brainfuck', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_brainfuck).default;
  }
});

var _cal = require('./cal');

Object.defineProperty(exports, 'cal', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_cal).default;
  }
});

var _capnproto = require('./capnproto');

Object.defineProperty(exports, 'capnproto', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_capnproto).default;
  }
});

var _ceylon = require('./ceylon');

Object.defineProperty(exports, 'ceylon', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ceylon).default;
  }
});

var _clean = require('./clean');

Object.defineProperty(exports, 'clean', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_clean).default;
  }
});

var _clojureRepl = require('./clojure-repl');

Object.defineProperty(exports, 'clojureRepl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_clojureRepl).default;
  }
});

var _clojure = require('./clojure');

Object.defineProperty(exports, 'clojure', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_clojure).default;
  }
});

var _cmake = require('./cmake');

Object.defineProperty(exports, 'cmake', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_cmake).default;
  }
});

var _coffeescript = require('./coffeescript');

Object.defineProperty(exports, 'coffeescript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_coffeescript).default;
  }
});

var _coq = require('./coq');

Object.defineProperty(exports, 'coq', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_coq).default;
  }
});

var _cos = require('./cos');

Object.defineProperty(exports, 'cos', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_cos).default;
  }
});

var _cpp = require('./cpp');

Object.defineProperty(exports, 'cpp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_cpp).default;
  }
});

var _crmsh = require('./crmsh');

Object.defineProperty(exports, 'crmsh', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_crmsh).default;
  }
});

var _crystal = require('./crystal');

Object.defineProperty(exports, 'crystal', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_crystal).default;
  }
});

var _cs = require('./cs');

Object.defineProperty(exports, 'cs', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_cs).default;
  }
});

var _csp = require('./csp');

Object.defineProperty(exports, 'csp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_csp).default;
  }
});

var _css = require('./css');

Object.defineProperty(exports, 'css', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_css).default;
  }
});

var _d = require('./d');

Object.defineProperty(exports, 'd', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_d).default;
  }
});

var _dart = require('./dart');

Object.defineProperty(exports, 'dart', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dart).default;
  }
});

var _delphi = require('./delphi');

Object.defineProperty(exports, 'delphi', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_delphi).default;
  }
});

var _diff = require('./diff');

Object.defineProperty(exports, 'diff', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_diff).default;
  }
});

var _django = require('./django');

Object.defineProperty(exports, 'django', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_django).default;
  }
});

var _dns = require('./dns');

Object.defineProperty(exports, 'dns', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dns).default;
  }
});

var _dockerfile = require('./dockerfile');

Object.defineProperty(exports, 'dockerfile', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dockerfile).default;
  }
});

var _dos = require('./dos');

Object.defineProperty(exports, 'dos', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dos).default;
  }
});

var _dsconfig = require('./dsconfig');

Object.defineProperty(exports, 'dsconfig', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dsconfig).default;
  }
});

var _dts = require('./dts');

Object.defineProperty(exports, 'dts', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dts).default;
  }
});

var _dust = require('./dust');

Object.defineProperty(exports, 'dust', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dust).default;
  }
});

var _ebnf = require('./ebnf');

Object.defineProperty(exports, 'ebnf', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ebnf).default;
  }
});

var _elixir = require('./elixir');

Object.defineProperty(exports, 'elixir', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_elixir).default;
  }
});

var _elm = require('./elm');

Object.defineProperty(exports, 'elm', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_elm).default;
  }
});

var _erb = require('./erb');

Object.defineProperty(exports, 'erb', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_erb).default;
  }
});

var _erlangRepl = require('./erlang-repl');

Object.defineProperty(exports, 'erlangRepl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_erlangRepl).default;
  }
});

var _erlang = require('./erlang');

Object.defineProperty(exports, 'erlang', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_erlang).default;
  }
});

var _excel = require('./excel');

Object.defineProperty(exports, 'excel', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_excel).default;
  }
});

var _fix = require('./fix');

Object.defineProperty(exports, 'fix', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_fix).default;
  }
});

var _flix = require('./flix');

Object.defineProperty(exports, 'flix', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_flix).default;
  }
});

var _fortran = require('./fortran');

Object.defineProperty(exports, 'fortran', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_fortran).default;
  }
});

var _fsharp = require('./fsharp');

Object.defineProperty(exports, 'fsharp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_fsharp).default;
  }
});

var _gams = require('./gams');

Object.defineProperty(exports, 'gams', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gams).default;
  }
});

var _gauss = require('./gauss');

Object.defineProperty(exports, 'gauss', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gauss).default;
  }
});

var _gcode = require('./gcode');

Object.defineProperty(exports, 'gcode', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gcode).default;
  }
});

var _gherkin = require('./gherkin');

Object.defineProperty(exports, 'gherkin', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gherkin).default;
  }
});

var _glsl = require('./glsl');

Object.defineProperty(exports, 'glsl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_glsl).default;
  }
});

var _go = require('./go');

Object.defineProperty(exports, 'go', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_go).default;
  }
});

var _golo = require('./golo');

Object.defineProperty(exports, 'golo', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_golo).default;
  }
});

var _gradle = require('./gradle');

Object.defineProperty(exports, 'gradle', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_gradle).default;
  }
});

var _groovy = require('./groovy');

Object.defineProperty(exports, 'groovy', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_groovy).default;
  }
});

var _haml = require('./haml');

Object.defineProperty(exports, 'haml', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_haml).default;
  }
});

var _handlebars = require('./handlebars');

Object.defineProperty(exports, 'handlebars', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_handlebars).default;
  }
});

var _haskell = require('./haskell');

Object.defineProperty(exports, 'haskell', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_haskell).default;
  }
});

var _haxe = require('./haxe');

Object.defineProperty(exports, 'haxe', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_haxe).default;
  }
});

var _hsp = require('./hsp');

Object.defineProperty(exports, 'hsp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_hsp).default;
  }
});

var _htmlbars = require('./htmlbars');

Object.defineProperty(exports, 'htmlbars', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_htmlbars).default;
  }
});

var _http = require('./http');

Object.defineProperty(exports, 'http', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_http).default;
  }
});

var _hy = require('./hy');

Object.defineProperty(exports, 'hy', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_hy).default;
  }
});

var _inform = require('./inform7');

Object.defineProperty(exports, 'inform7', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_inform).default;
  }
});

var _ini = require('./ini');

Object.defineProperty(exports, 'ini', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ini).default;
  }
});

var _irpf = require('./irpf90');

Object.defineProperty(exports, 'irpf90', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_irpf).default;
  }
});

var _java = require('./java');

Object.defineProperty(exports, 'java', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_java).default;
  }
});

var _javascript = require('./javascript');

Object.defineProperty(exports, 'javascript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_javascript).default;
  }
});

var _jbossCli = require('./jboss-cli');

Object.defineProperty(exports, 'jbossCli', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_jbossCli).default;
  }
});

var _json = require('./json');

Object.defineProperty(exports, 'json', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_json).default;
  }
});

var _juliaRepl = require('./julia-repl');

Object.defineProperty(exports, 'juliaRepl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_juliaRepl).default;
  }
});

var _julia = require('./julia');

Object.defineProperty(exports, 'julia', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_julia).default;
  }
});

var _kotlin = require('./kotlin');

Object.defineProperty(exports, 'kotlin', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_kotlin).default;
  }
});

var _lasso = require('./lasso');

Object.defineProperty(exports, 'lasso', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_lasso).default;
  }
});

var _ldif = require('./ldif');

Object.defineProperty(exports, 'ldif', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ldif).default;
  }
});

var _leaf = require('./leaf');

Object.defineProperty(exports, 'leaf', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_leaf).default;
  }
});

var _less = require('./less');

Object.defineProperty(exports, 'less', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_less).default;
  }
});

var _lisp = require('./lisp');

Object.defineProperty(exports, 'lisp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_lisp).default;
  }
});

var _livecodeserver = require('./livecodeserver');

Object.defineProperty(exports, 'livecodeserver', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_livecodeserver).default;
  }
});

var _livescript = require('./livescript');

Object.defineProperty(exports, 'livescript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_livescript).default;
  }
});

var _llvm = require('./llvm');

Object.defineProperty(exports, 'llvm', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_llvm).default;
  }
});

var _lsl = require('./lsl');

Object.defineProperty(exports, 'lsl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_lsl).default;
  }
});

var _lua = require('./lua');

Object.defineProperty(exports, 'lua', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_lua).default;
  }
});

var _makefile = require('./makefile');

Object.defineProperty(exports, 'makefile', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_makefile).default;
  }
});

var _markdown = require('./markdown');

Object.defineProperty(exports, 'markdown', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_markdown).default;
  }
});

var _mathematica = require('./mathematica');

Object.defineProperty(exports, 'mathematica', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mathematica).default;
  }
});

var _matlab = require('./matlab');

Object.defineProperty(exports, 'matlab', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_matlab).default;
  }
});

var _maxima = require('./maxima');

Object.defineProperty(exports, 'maxima', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_maxima).default;
  }
});

var _mel = require('./mel');

Object.defineProperty(exports, 'mel', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mel).default;
  }
});

var _mercury = require('./mercury');

Object.defineProperty(exports, 'mercury', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mercury).default;
  }
});

var _mipsasm = require('./mipsasm');

Object.defineProperty(exports, 'mipsasm', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mipsasm).default;
  }
});

var _mizar = require('./mizar');

Object.defineProperty(exports, 'mizar', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mizar).default;
  }
});

var _mojolicious = require('./mojolicious');

Object.defineProperty(exports, 'mojolicious', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mojolicious).default;
  }
});

var _monkey = require('./monkey');

Object.defineProperty(exports, 'monkey', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_monkey).default;
  }
});

var _moonscript = require('./moonscript');

Object.defineProperty(exports, 'moonscript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_moonscript).default;
  }
});

var _n1ql = require('./n1ql');

Object.defineProperty(exports, 'n1ql', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_n1ql).default;
  }
});

var _nginx = require('./nginx');

Object.defineProperty(exports, 'nginx', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_nginx).default;
  }
});

var _nimrod = require('./nimrod');

Object.defineProperty(exports, 'nimrod', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_nimrod).default;
  }
});

var _nix = require('./nix');

Object.defineProperty(exports, 'nix', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_nix).default;
  }
});

var _nsis = require('./nsis');

Object.defineProperty(exports, 'nsis', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_nsis).default;
  }
});

var _objectivec = require('./objectivec');

Object.defineProperty(exports, 'objectivec', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_objectivec).default;
  }
});

var _ocaml = require('./ocaml');

Object.defineProperty(exports, 'ocaml', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ocaml).default;
  }
});

var _openscad = require('./openscad');

Object.defineProperty(exports, 'openscad', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_openscad).default;
  }
});

var _oxygene = require('./oxygene');

Object.defineProperty(exports, 'oxygene', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_oxygene).default;
  }
});

var _parser = require('./parser3');

Object.defineProperty(exports, 'parser3', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_parser).default;
  }
});

var _perl = require('./perl');

Object.defineProperty(exports, 'perl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_perl).default;
  }
});

var _pf = require('./pf');

Object.defineProperty(exports, 'pf', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_pf).default;
  }
});

var _php = require('./php');

Object.defineProperty(exports, 'php', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_php).default;
  }
});

var _pony = require('./pony');

Object.defineProperty(exports, 'pony', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_pony).default;
  }
});

var _powershell = require('./powershell');

Object.defineProperty(exports, 'powershell', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_powershell).default;
  }
});

var _processing = require('./processing');

Object.defineProperty(exports, 'processing', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_processing).default;
  }
});

var _profile = require('./profile');

Object.defineProperty(exports, 'profile', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_profile).default;
  }
});

var _prolog = require('./prolog');

Object.defineProperty(exports, 'prolog', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_prolog).default;
  }
});

var _protobuf = require('./protobuf');

Object.defineProperty(exports, 'protobuf', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_protobuf).default;
  }
});

var _puppet = require('./puppet');

Object.defineProperty(exports, 'puppet', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_puppet).default;
  }
});

var _purebasic = require('./purebasic');

Object.defineProperty(exports, 'purebasic', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_purebasic).default;
  }
});

var _python = require('./python');

Object.defineProperty(exports, 'python', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_python).default;
  }
});

var _q = require('./q');

Object.defineProperty(exports, 'q', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_q).default;
  }
});

var _qml = require('./qml');

Object.defineProperty(exports, 'qml', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_qml).default;
  }
});

var _r = require('./r');

Object.defineProperty(exports, 'r', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_r).default;
  }
});

var _rib = require('./rib');

Object.defineProperty(exports, 'rib', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_rib).default;
  }
});

var _roboconf = require('./roboconf');

Object.defineProperty(exports, 'roboconf', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_roboconf).default;
  }
});

var _routeros = require('./routeros');

Object.defineProperty(exports, 'routeros', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_routeros).default;
  }
});

var _rsl = require('./rsl');

Object.defineProperty(exports, 'rsl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_rsl).default;
  }
});

var _ruby = require('./ruby');

Object.defineProperty(exports, 'ruby', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ruby).default;
  }
});

var _ruleslanguage = require('./ruleslanguage');

Object.defineProperty(exports, 'ruleslanguage', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ruleslanguage).default;
  }
});

var _rust = require('./rust');

Object.defineProperty(exports, 'rust', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_rust).default;
  }
});

var _scala = require('./scala');

Object.defineProperty(exports, 'scala', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scala).default;
  }
});

var _scheme = require('./scheme');

Object.defineProperty(exports, 'scheme', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scheme).default;
  }
});

var _scilab = require('./scilab');

Object.defineProperty(exports, 'scilab', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scilab).default;
  }
});

var _scss = require('./scss');

Object.defineProperty(exports, 'scss', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_scss).default;
  }
});

var _shell = require('./shell');

Object.defineProperty(exports, 'shell', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_shell).default;
  }
});

var _smali = require('./smali');

Object.defineProperty(exports, 'smali', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_smali).default;
  }
});

var _smalltalk = require('./smalltalk');

Object.defineProperty(exports, 'smalltalk', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_smalltalk).default;
  }
});

var _sml = require('./sml');

Object.defineProperty(exports, 'sml', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sml).default;
  }
});

var _sqf = require('./sqf');

Object.defineProperty(exports, 'sqf', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sqf).default;
  }
});

var _sql = require('./sql');

Object.defineProperty(exports, 'sql', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sql).default;
  }
});

var _stan = require('./stan');

Object.defineProperty(exports, 'stan', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_stan).default;
  }
});

var _stata = require('./stata');

Object.defineProperty(exports, 'stata', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_stata).default;
  }
});

var _step = require('./step21');

Object.defineProperty(exports, 'step21', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_step).default;
  }
});

var _stylus = require('./stylus');

Object.defineProperty(exports, 'stylus', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_stylus).default;
  }
});

var _subunit = require('./subunit');

Object.defineProperty(exports, 'subunit', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_subunit).default;
  }
});

var _swift = require('./swift');

Object.defineProperty(exports, 'swift', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_swift).default;
  }
});

var _taggerscript = require('./taggerscript');

Object.defineProperty(exports, 'taggerscript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_taggerscript).default;
  }
});

var _tap = require('./tap');

Object.defineProperty(exports, 'tap', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_tap).default;
  }
});

var _tcl = require('./tcl');

Object.defineProperty(exports, 'tcl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_tcl).default;
  }
});

var _tex = require('./tex');

Object.defineProperty(exports, 'tex', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_tex).default;
  }
});

var _thrift = require('./thrift');

Object.defineProperty(exports, 'thrift', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_thrift).default;
  }
});

var _tp = require('./tp');

Object.defineProperty(exports, 'tp', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_tp).default;
  }
});

var _twig = require('./twig');

Object.defineProperty(exports, 'twig', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_twig).default;
  }
});

var _typescript = require('./typescript');

Object.defineProperty(exports, 'typescript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_typescript).default;
  }
});

var _vala = require('./vala');

Object.defineProperty(exports, 'vala', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vala).default;
  }
});

var _vbnet = require('./vbnet');

Object.defineProperty(exports, 'vbnet', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vbnet).default;
  }
});

var _vbscriptHtml = require('./vbscript-html');

Object.defineProperty(exports, 'vbscriptHtml', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vbscriptHtml).default;
  }
});

var _vbscript = require('./vbscript');

Object.defineProperty(exports, 'vbscript', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vbscript).default;
  }
});

var _verilog = require('./verilog');

Object.defineProperty(exports, 'verilog', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_verilog).default;
  }
});

var _vhdl = require('./vhdl');

Object.defineProperty(exports, 'vhdl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vhdl).default;
  }
});

var _vim = require('./vim');

Object.defineProperty(exports, 'vim', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vim).default;
  }
});

var _x86asm = require('./x86asm');

Object.defineProperty(exports, 'x86asm', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_x86asm).default;
  }
});

var _xl = require('./xl');

Object.defineProperty(exports, 'xl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_xl).default;
  }
});

var _xml = require('./xml');

Object.defineProperty(exports, 'xml', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_xml).default;
  }
});

var _xquery = require('./xquery');

Object.defineProperty(exports, 'xquery', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_xquery).default;
  }
});

var _yaml = require('./yaml');

Object.defineProperty(exports, 'yaml', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_yaml).default;
  }
});

var _zephir = require('./zephir');

Object.defineProperty(exports, 'zephir', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_zephir).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
