// encodable/esm/fillers/completeChannelDef.js
"use strict";

exports.__esModule = true;
exports.default = completeChannelDef;

var _ChannelDef = require("../typeGuards/ChannelDef");

var _completeAxisConfig = _interopRequireDefault(require("./completeAxisConfig"));

var _completeLegendConfig = _interopRequireDefault(require("./completeLegendConfig"));

var _completeScaleConfig = _interopRequireDefault(require("./completeScaleConfig"));

var _inferFieldType = _interopRequireDefault(require("./inferFieldType"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function completeChannelDef(channelType, channelDef) {
  var _channelDef$title;

  if ((0, _ChannelDef.isValueDef)(channelDef)) {
    return { ...channelDef,
      axis: false,
      legend: false,
      scale: false,
      title: ''
    };
  } // Fill top-level properties


  const copy = { ...channelDef,
    title: (0, _ChannelDef.isFieldDef)(channelDef) ? (_channelDef$title = channelDef.title) != null ? _channelDef$title : channelDef.field : '',
    type: (0, _ChannelDef.isTypedFieldDef)(channelDef) ? channelDef.type : (0, _inferFieldType.default)(channelType, channelDef.field)
  }; // Scale needs the top-level properties to be filled.

  const scale = (0, _completeScaleConfig.default)(channelType, copy);
  const copy2 = { ...copy,
    scale
  }; // These two rely on scale

  const axis = (0, _completeAxisConfig.default)(channelType, copy2);
  const legend = (0, _completeLegendConfig.default)(channelType, copy2);
  return { ...copy2,
    axis,
    legend
  };
}
