// encodable/lib/types/VegaLite.js
"use strict";

exports.__esModule = true;
exports.ScaleType = exports.Type = exports.Axis = exports.NiceTime = exports.Scale = exports.SchemeParams = exports.DateTime = exports.isDateTime = exports.Value = exports.ValueDef = void 0;

var _channeldef = require("vega-lite/build/src/channeldef");

exports.ValueDef = _channeldef.ValueDef;
exports.Value = _channeldef.Value;

var _datetime = require("vega-lite/build/src/datetime");

exports.isDateTime = _datetime.isDateTime;
exports.DateTime = _datetime.DateTime;

var _scale = require("vega-lite/build/src/scale");

exports.SchemeParams = _scale.SchemeParams;
exports.Scale = _scale.Scale;
exports.NiceTime = _scale.NiceTime;

var _axis = require("vega-lite/build/src/axis");

exports.Axis = _axis.Axis;

var _type = require("vega-lite/build/src/type");

exports.Type = _type.Type;
// Types imported from vega-lite
// Override this implementation
// because vega-lite uses namespace which has issues with babel and typescript
const ScaleType = {
  // Continuous - Quantitative
  LINEAR: 'linear',
  LOG: 'log',
  POW: 'pow',
  SQRT: 'sqrt',
  SYMLOG: 'symlog',
  // Continuous - Time
  TIME: 'time',
  UTC: 'utc',
  // Discretizing scales
  QUANTILE: 'quantile',
  QUANTIZE: 'quantize',
  THRESHOLD: 'threshold',
  BIN_ORDINAL: 'bin-ordinal',
  // Discrete scales
  ORDINAL: 'ordinal',
  POINT: 'point',
  BAND: 'band'
};
exports.ScaleType = ScaleType;
