// react-syntax-highlighter/styles/prism/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _coy = require('./coy');

Object.defineProperty(exports, 'coy', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_coy).default;
  }
});

var _dark = require('./dark');

Object.defineProperty(exports, 'dark', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_dark).default;
  }
});

var _funky = require('./funky');

Object.defineProperty(exports, 'funky', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_funky).default;
  }
});

var _okaidia = require('./okaidia');

Object.defineProperty(exports, 'okaidia', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_okaidia).default;
  }
});

var _solarizedlight = require('./solarizedlight');

Object.defineProperty(exports, 'solarizedlight', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_solarizedlight).default;
  }
});

var _tomorrow = require('./tomorrow');

Object.defineProperty(exports, 'tomorrow', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_tomorrow).default;
  }
});

var _twilight = require('./twilight');

Object.defineProperty(exports, 'twilight', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_twilight).default;
  }
});

var _prism = require('./prism');

Object.defineProperty(exports, 'prism', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_prism).default;
  }
});

var _atomDark = require('./atom-dark');

Object.defineProperty(exports, 'atomDark', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_atomDark).default;
  }
});

var _base16Ateliersulphurpool = require('./base16-ateliersulphurpool.light');

Object.defineProperty(exports, 'base16AteliersulphurpoolLight', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_base16Ateliersulphurpool).default;
  }
});

var _cb = require('./cb');

Object.defineProperty(exports, 'cb', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_cb).default;
  }
});

var _darcula = require('./darcula');

Object.defineProperty(exports, 'darcula', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_darcula).default;
  }
});

var _duotoneDark = require('./duotone-dark');

Object.defineProperty(exports, 'duotoneDark', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_duotoneDark).default;
  }
});

var _duotoneEarth = require('./duotone-earth');

Object.defineProperty(exports, 'duotoneEarth', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_duotoneEarth).default;
  }
});

var _duotoneForest = require('./duotone-forest');

Object.defineProperty(exports, 'duotoneForest', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_duotoneForest).default;
  }
});

var _duotoneLight = require('./duotone-light');

Object.defineProperty(exports, 'duotoneLight', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_duotoneLight).default;
  }
});

var _duotoneSea = require('./duotone-sea');

Object.defineProperty(exports, 'duotoneSea', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_duotoneSea).default;
  }
});

var _duotoneSpace = require('./duotone-space');

Object.defineProperty(exports, 'duotoneSpace', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_duotoneSpace).default;
  }
});

var _ghcolors = require('./ghcolors');

Object.defineProperty(exports, 'ghcolors', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_ghcolors).default;
  }
});

var _hopscotch = require('./hopscotch');

Object.defineProperty(exports, 'hopscotch', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_hopscotch).default;
  }
});

var _pojoaque = require('./pojoaque');

Object.defineProperty(exports, 'pojoaque', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_pojoaque).default;
  }
});

var _vs = require('./vs');

Object.defineProperty(exports, 'vs', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_vs).default;
  }
});

var _xonokai = require('./xonokai');

Object.defineProperty(exports, 'xonokai', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_xonokai).default;
  }
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
