// @superset-ui/legacy-plugin-chart-sankey-loop/lib/SankeyLoop.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _d3Selection = require("d3-selection");

var _d3SankeyDiagram = require("d3-sankey-diagram");

var _color = require("@superset-ui/color");

var _numberFormat = require("@superset-ui/number-format");

require("./SankeyLoop.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

// a problem with 'd3-sankey-diagram'  is that the sankey().extent() paramters, which
// informs the layout of the bounding box of the sankey columns, does not account
// for labels and paths which happen to be layedout outside that rectangle.
// for that reason i've selected relatively large default left/right margins, and have
// made 'margin' a property.   i have raised an issue in the chart repo:
//
//   https://github.com/ricklupton/d3-sankey-diagram/issues/20
const defaultMargin = {
  top: 0,
  right: 80,
  bottom: 0,
  left: 80
};
const propTypes = {
  data: _propTypes.default.arrayOf(_propTypes.default.shape({
    source: _propTypes.default.string,
    target: _propTypes.default.string,
    value: _propTypes.default.number
  })),
  width: _propTypes.default.number,
  height: _propTypes.default.number,
  colorScheme: _propTypes.default.string,
  margin: _propTypes.default.shape({
    top: _propTypes.default.number,
    right: _propTypes.default.number,
    bottom: _propTypes.default.number,
    left: _propTypes.default.number
  })
};
const percentFormat = (0, _numberFormat.getNumberFormatter)(_numberFormat.NumberFormats.PERCENT_1_POINT);
const countFormat = (0, _numberFormat.getNumberFormatter)();

function computeGraph(links) {
  // this assumes source and target are string values
  const nodes = Array.from(links.reduce((set, {
    source,
    target
  }) => set.add(source).add(target), new Set())).map(id => ({
    id,
    name: id
  }));
  return {
    nodes,
    // links are shallow copied as the chart layout modifies them, and it is best to
    // leave the passed data un-altered
    links: links.map(d => _extends({}, d))
  };
}

function SankeyLoop(element, props) {
  const {
    data,
    width,
    height,
    colorScheme
  } = props;

  const color = _color.CategoricalColorNamespace.getScale(colorScheme);

  const margin = _extends({}, defaultMargin, {}, props.margin);

  const innerWidth = width - margin.left - margin.right;
  const innerHeight = height - margin.top - margin.bottom;
  const layout = (0, _d3SankeyDiagram.sankey)().nodeId(d => d.id).extent([[margin.left, margin.top], [innerWidth, innerHeight]]);
  const diagram = (0, _d3SankeyDiagram.sankeyDiagram)().nodeTitle(d => d.name).linkTitle(({
    source: {
      name: sName,
      value: sValue
    },
    target: {
      name: tName
    },
    value
  }) => sName + " \u2192 " + tName + ": " + countFormat(value) + " (" + percentFormat(value / sValue) + ")").linkColor(d => color(d.source.name));
  const svg = (0, _d3Selection.select)(element).append('svg').classed('superset-legacy-chart-sankey-loop', true).style('width', width).style('height', height).datum(layout(computeGraph(data))).call(diagram);
  svg.selectAll('g.link').classed('link', true).append('text').attr('x', d => d.points[0].x).attr('y', d => d.points[0].y).attr('dy', 3).attr('dx', 2).text(d => countFormat(d.value) + " (" + percentFormat(d.value / d.source.value) + ")");
}

SankeyLoop.displayName = 'SankeyLoop';
SankeyLoop.propTypes = propTypes;
var _default = SankeyLoop;
exports.default = _default;
