// @superset-ui/connection/lib/SupersetClient.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _SupersetClientClass = _interopRequireDefault(require("./SupersetClientClass"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

let singletonClient;

function getInstance() {
  if (!singletonClient) {
    throw new Error('You must call SupersetClient.configure(...) before calling other methods');
  }

  return singletonClient;
}

const SupersetClient = {
  configure: config => {
    singletonClient = new _SupersetClientClass.default(config);
    return singletonClient;
  },
  delete: request => getInstance().delete(request),
  get: request => getInstance().get(request),
  getInstance,
  init: force => getInstance().init(force),
  isAuthenticated: () => getInstance().isAuthenticated(),
  post: request => getInstance().post(request),
  put: request => getInstance().put(request),
  reAuthenticate: () => getInstance().init(
  /* force = */
  true),
  request: _request => getInstance().request(_request),
  reset: () => {
    singletonClient = undefined;
  }
};
var _default = SupersetClient;
exports.default = _default;
