// @superset-ui/legacy-plugin-chart-iframe/lib/Iframe.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _mustache = _interopRequireDefault(require("mustache"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const propTypes = {
  className: _propTypes.default.string,
  height: _propTypes.default.number.isRequired,
  url: _propTypes.default.string,
  width: _propTypes.default.number.isRequired
};
const defaultProps = {
  className: '',
  url: ''
};

class Iframe extends _react.default.PureComponent {
  render() {
    const {
      className,
      url,
      width,
      height
    } = this.props;

    const completeUrl = _mustache.default.render(url, {
      height,
      width
    });

    return /*#__PURE__*/_react.default.createElement("iframe", {
      className: className,
      title: "superset-iframe",
      src: completeUrl,
      style: {
        height,
        width: '100%'
      }
    });
  }

}

Iframe.propTypes = propTypes;
Iframe.defaultProps = defaultProps;
var _default = Iframe;
exports.default = _default;
