// @superset-ui/legacy-plugin-chart-event-flow/lib/EventFlow.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _eventFlow = require("@data-ui/event-flow");

var _translation = require("@superset-ui/translation");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable react/forbid-prop-types */
const propTypes = {
  data: _propTypes.default.array,
  height: _propTypes.default.number,

  /* eslint-disable-next-line */
  initialMinEventCount: _propTypes.default.number,
  width: _propTypes.default.number
};
const defaultProps = {
  data: null,
  height: 400,
  width: 400
};

function CustomEventFlow(props) {
  const {
    data,
    height,
    initialMinEventCount,
    width
  } = props;

  if (data) {
    return /*#__PURE__*/_react.default.createElement(_eventFlow.App, {
      width: width,
      height: height,
      data: data,
      initialMinEventCount: initialMinEventCount,
      initialShowControls: false
    });
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      height,
      width
    }
  }, /*#__PURE__*/_react.default.createElement("div", null, (0, _translation.t)('Sorry, there appears to be no data')));
}

CustomEventFlow.propTypes = propTypes;
CustomEventFlow.defaultProps = defaultProps;
var _default = CustomEventFlow;
exports.default = _default;
