// @superset-ui/legacy-preset-chart-deckgl/lib/components/Legend.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _numberFormat = require("@superset-ui/number-format");

require("./Legend.css");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* eslint-disable react/jsx-sort-default-props */

/* eslint-disable react/sort-prop-types */

/* eslint-disable jsx-a11y/anchor-is-valid */

/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
const categoryDelimiter = ' - ';
const propTypes = {
  categories: _propTypes.default.object,
  toggleCategory: _propTypes.default.func,
  showSingleCategory: _propTypes.default.func,
  format: _propTypes.default.string,
  position: _propTypes.default.oneOf([null, 'tl', 'tr', 'bl', 'br'])
};
const defaultProps = {
  categories: {},
  toggleCategory: () => {},
  showSingleCategory: () => {},
  format: null,
  position: 'tr'
};

class Legend extends _react.default.PureComponent {
  format(value) {
    if (!this.props.format) {
      return value;
    }

    const numValue = parseFloat(value);
    return (0, _numberFormat.formatNumber)(this.props.format, numValue);
  }

  formatCategoryLabel(k) {
    if (!this.props.format) {
      return k;
    }

    if (k.includes(categoryDelimiter)) {
      const values = k.split(categoryDelimiter);
      return this.format(values[0]) + categoryDelimiter + this.format(values[1]);
    }

    return this.format(k);
  }

  render() {
    if (Object.keys(this.props.categories).length === 0 || this.props.position === null) {
      return null;
    }

    const categories = Object.entries(this.props.categories).map((_ref) => {
      let [k, v] = _ref;
      const style = {
        color: "rgba(" + v.color.join(', ') + ")"
      };
      const icon = v.enabled ? '\u25FC' : '\u25FB';
      return _react.default.createElement("li", {
        key: k
      }, _react.default.createElement("a", {
        href: "#",
        onClick: () => this.props.toggleCategory(k),
        onDoubleClick: () => this.props.showSingleCategory(k)
      }, _react.default.createElement("span", {
        style: style
      }, icon), " ", this.formatCategoryLabel(k)));
    });
    const vertical = this.props.position.charAt(0) === 't' ? 'top' : 'bottom';
    const horizontal = this.props.position.charAt(1) === 'r' ? 'right' : 'left';
    const style = {
      position: 'absolute',
      [vertical]: '0px',
      [horizontal]: '10px'
    };
    return _react.default.createElement("div", {
      className: "legend",
      style: style
    }, _react.default.createElement("ul", {
      className: "categories"
    }, categories));
  }

}

exports.default = Legend;
Legend.propTypes = propTypes;
Legend.defaultProps = defaultProps;
