// @data-ui/xy-chart/lib/series/LineSeries.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _theme = require("@data-ui/theme");

var _shared = require("@data-ui/shared");

var _glyph = require("@vx/glyph");

var _shape = require("@vx/shape");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _chartUtils = require("../utils/chartUtils");

var _findClosestDatum = _interopRequireDefault(require("../utils/findClosestDatum"));

var _interpolatorLookup = _interopRequireDefault(require("../utils/interpolatorLookup"));

var _propShapes = require("../utils/propShapes");

var _sharedSeriesProps = _interopRequireDefault(require("../utils/sharedSeriesProps"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _inheritsLoose(subClass, superClass) { subClass.prototype = Object.create(superClass.prototype); subClass.prototype.constructor = subClass; subClass.__proto__ = superClass; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var propTypes = _extends({}, _sharedSeriesProps.default, {
  data: _propShapes.lineSeriesDataShape.isRequired,
  interpolation: _propShapes.interpolationShape,
  showPoints: _propTypes.default.bool,
  stroke: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  strokeDasharray: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.string]),
  strokeWidth: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.number]),
  strokeOpacity: _propTypes.default.oneOfType([_propTypes.default.func, _propTypes.default.number]),
  strokeLinecap: _propTypes.default.oneOf(['butt', 'square', 'round', 'inherit'])
});

var defaultProps = {
  interpolation: 'monotoneX',
  showPoints: false,
  stroke: _theme.color.default,
  strokeDasharray: null,
  strokeWidth: 3,
  strokeOpacity: 1,
  strokeLinecap: 'round'
};

var x = function x(d) {
  return d.x;
};

var y = function y(d) {
  return d.y;
};

var defined = function defined(d) {
  return (0, _chartUtils.isDefined)(y(d));
};

var noEventsStyles = {
  pointerEvents: 'none'
};

var LineSeries =
/*#__PURE__*/
function (_React$PureComponent) {
  _inheritsLoose(LineSeries, _React$PureComponent);

  function LineSeries() {
    return _React$PureComponent.apply(this, arguments) || this;
  }

  var _proto = LineSeries.prototype;

  _proto.render = function render() {
    var _this$props = this.props,
        data = _this$props.data,
        disableMouseEvents = _this$props.disableMouseEvents,
        interpolation = _this$props.interpolation,
        showPoints = _this$props.showPoints,
        stroke = _this$props.stroke,
        strokeDasharray = _this$props.strokeDasharray,
        strokeWidth = _this$props.strokeWidth,
        strokeLinecap = _this$props.strokeLinecap,
        strokeOpacity = _this$props.strokeOpacity,
        xScale = _this$props.xScale,
        yScale = _this$props.yScale,
        margin = _this$props.margin,
        onClick = _this$props.onClick,
        onMouseMove = _this$props.onMouseMove,
        onMouseLeave = _this$props.onMouseLeave;
    if (!xScale || !yScale) return null;
    var strokeValue = (0, _chartUtils.callOrValue)(stroke);
    var strokeOpacityValue = (0, _chartUtils.callOrValue)(strokeOpacity);
    var curve = _interpolatorLookup.default[interpolation] || _interpolatorLookup.default.monotoneX;
    return _react.default.createElement(_shape.LinePath, {
      style: disableMouseEvents ? noEventsStyles : null,
      data: data,
      xScale: xScale,
      yScale: yScale,
      x: x,
      y: y,
      stroke: strokeValue,
      strokeWidth: (0, _chartUtils.callOrValue)(strokeWidth),
      strokeDasharray: (0, _chartUtils.callOrValue)(strokeDasharray),
      strokeLinecap: strokeLinecap,
      strokeOpacity: strokeOpacityValue,
      curve: curve,
      defined: defined,
      onClick: disableMouseEvents ? null : onClick && function () {
        return function (event) {
          var d = (0, _findClosestDatum.default)({
            data: data,
            getX: x,
            event: event,
            xScale: xScale,
            marginLeft: margin.left
          });
          onClick({
            event: event,
            data: data,
            datum: d,
            color: strokeValue
          });
        };
      },
      onMouseMove: disableMouseEvents ? null : onMouseMove && function () {
        return function (event) {
          var d = (0, _findClosestDatum.default)({
            data: data,
            getX: x,
            event: event,
            xScale: xScale,
            marginLeft: margin.left
          });
          onMouseMove({
            event: event,
            data: data,
            datum: d,
            color: strokeValue
          });
        };
      },
      onMouseLeave: disableMouseEvents ? null : onMouseLeave && function () {
        return onMouseLeave;
      },
      glyph: function glyph(d, i) {
        return (// <a /> wrapper is needed for focusing with SVG 1.2 regardless of whether we show a point
          (0, _chartUtils.isDefined)(x(d)) && (0, _chartUtils.isDefined)(y(d)) && _react.default.createElement(_shared.FocusBlurHandler, {
            key: "linepoint-" + i,
            onBlur: disableMouseEvents ? null : onMouseLeave,
            onFocus: disableMouseEvents ? null : function (event) {
              onMouseMove({
                event: event,
                data: data,
                datum: d,
                color: d.stroke || (0, _chartUtils.callOrValue)(stroke, d, i),
                index: i
              });
            }
          }, showPoints && _react.default.createElement(_glyph.GlyphDot, {
            key: i + "-" + x(d),
            cx: xScale(x(d)),
            cy: yScale(y(d)),
            r: 4,
            fill: d.stroke || (0, _chartUtils.callOrValue)(stroke, d, i),
            stroke: "#FFFFFF",
            strokeWidth: 1,
            style: {
              pointerEvents: 'none'
            }
          }, d.label && _react.default.createElement("text", {
            x: xScale(x(d)),
            y: yScale(y(d)),
            dx: 10,
            fill: d.stroke || (0, _chartUtils.callOrValue)(stroke, d, i),
            stroke: "#fff",
            strokeWidth: 1,
            fontSize: 12
          }, d.label)))
        );
      }
    });
  };

  return LineSeries;
}(_react.default.PureComponent);

exports.default = LineSeries;
LineSeries.propTypes = propTypes;
LineSeries.defaultProps = defaultProps;
LineSeries.displayName = 'LineSeries';
