// @vx/scale/lib/index.js
"use strict";

exports.__esModule = true;
exports.scaleSqrt = exports.updateScale = exports.scaleThreshold = exports.scaleSymlog = exports.scaleQuantile = exports.scaleQuantize = exports.scaleOrdinal = exports.scalePower = exports.scaleLog = exports.scaleUtc = exports.scaleTime = exports.scaleLinear = exports.scalePoint = exports.scaleBand = void 0;

var _band = _interopRequireDefault(require("./scales/band"));

exports.scaleBand = _band.default;

var _point = _interopRequireDefault(require("./scales/point"));

exports.scalePoint = _point.default;

var _linear = _interopRequireDefault(require("./scales/linear"));

exports.scaleLinear = _linear.default;

var _time = _interopRequireDefault(require("./scales/time"));

exports.scaleTime = _time.default;

var _utc = _interopRequireDefault(require("./scales/utc"));

exports.scaleUtc = _utc.default;

var _log = _interopRequireDefault(require("./scales/log"));

exports.scaleLog = _log.default;

var _power = _interopRequireDefault(require("./scales/power"));

exports.scalePower = _power.default;

var _ordinal = _interopRequireDefault(require("./scales/ordinal"));

exports.scaleOrdinal = _ordinal.default;

var _quantize = _interopRequireDefault(require("./scales/quantize"));

exports.scaleQuantize = _quantize.default;

var _quantile = _interopRequireDefault(require("./scales/quantile"));

exports.scaleQuantile = _quantile.default;

var _symlog = _interopRequireDefault(require("./scales/symlog"));

exports.scaleSymlog = _symlog.default;

var _threshold = _interopRequireDefault(require("./scales/threshold"));

exports.scaleThreshold = _threshold.default;

var _updateScale = _interopRequireDefault(require("./util/updateScale"));

exports.updateScale = _updateScale.default;

var _squareRoot = _interopRequireDefault(require("./scales/squareRoot"));

exports.scaleSqrt = _squareRoot.default;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
