// react-style-proptype/src/Style.flow.js
// @flow
// generated by scripts/getFlow

export type Style = {
  alignContent?: ?string | ?number,
  alignItems?: ?string | ?number,
  alignSelf?: ?string | ?number,
  all?: ?string | ?number,
  animation?: ?string | ?number,
  animationDelay?: ?string | ?number,
  animationDirection?: ?string | ?number,
  animationDuration?: ?string | ?number,
  animationFillMode?: ?string | ?number,
  animationIterationCount?: ?string | ?number,
  animationName?: ?string | ?number,
  animationPlayState?: ?string | ?number,
  animationTimingFunction?: ?string | ?number,
  backfaceVisibility?: ?string | ?number,
  background?: ?string | ?number,
  backgroundAttachment?: ?string | ?number,
  backgroundBlendMode?: ?string | ?number,
  backgroundClip?: ?string | ?number,
  backgroundColor?: ?string | ?number,
  backgroundImage?: ?string | ?number,
  backgroundOrigin?: ?string | ?number,
  backgroundPosition?: ?string | ?number,
  backgroundRepeat?: ?string | ?number,
  backgroundSize?: ?string | ?number,
  blockSize?: ?string | ?number,
  border?: ?string | ?number,
  borderBlockEnd?: ?string | ?number,
  borderBlockEndColor?: ?string | ?number,
  borderBlockEndStyle?: ?string | ?number,
  borderBlockEndWidth?: ?string | ?number,
  borderBlockStart?: ?string | ?number,
  borderBlockStartColor?: ?string | ?number,
  borderBlockStartStyle?: ?string | ?number,
  borderBlockStartWidth?: ?string | ?number,
  borderBottom?: ?string | ?number,
  borderBottomColor?: ?string | ?number,
  borderBottomLeftRadius?: ?string | ?number,
  borderBottomRightRadius?: ?string | ?number,
  borderBottomStyle?: ?string | ?number,
  borderBottomWidth?: ?string | ?number,
  borderCollapse?: ?string | ?number,
  borderColor?: ?string | ?number,
  borderImage?: ?string | ?number,
  borderImageOutset?: ?string | ?number,
  borderImageRepeat?: ?string | ?number,
  borderImageSlice?: ?string | ?number,
  borderImageSource?: ?string | ?number,
  borderImageWidth?: ?string | ?number,
  borderInlineEnd?: ?string | ?number,
  borderInlineEndColor?: ?string | ?number,
  borderInlineEndStyle?: ?string | ?number,
  borderInlineEndWidth?: ?string | ?number,
  borderInlineStart?: ?string | ?number,
  borderInlineStartColor?: ?string | ?number,
  borderInlineStartStyle?: ?string | ?number,
  borderInlineStartWidth?: ?string | ?number,
  borderLeft?: ?string | ?number,
  borderLeftColor?: ?string | ?number,
  borderLeftStyle?: ?string | ?number,
  borderLeftWidth?: ?string | ?number,
  borderRadius?: ?string | ?number,
  borderRight?: ?string | ?number,
  borderRightColor?: ?string | ?number,
  borderRightStyle?: ?string | ?number,
  borderRightWidth?: ?string | ?number,
  borderSpacing?: ?string | ?number,
  borderStyle?: ?string | ?number,
  borderTop?: ?string | ?number,
  borderTopColor?: ?string | ?number,
  borderTopLeftRadius?: ?string | ?number,
  borderTopRightRadius?: ?string | ?number,
  borderTopStyle?: ?string | ?number,
  borderTopWidth?: ?string | ?number,
  borderWidth?: ?string | ?number,
  bottom?: ?string | ?number,
  boxDecorationBreak?: ?string | ?number,
  boxShadow?: ?string | ?number,
  boxSizing?: ?string | ?number,
  breakAfter?: ?string | ?number,
  breakBefore?: ?string | ?number,
  breakInside?: ?string | ?number,
  captionSide?: ?string | ?number,
  caretColor?: ?string | ?number,
  ch?: ?string | ?number,
  clear?: ?string | ?number,
  clip?: ?string | ?number,
  clipPath?: ?string | ?number,
  cm?: ?string | ?number,
  color?: ?string | ?number,
  columnCount?: ?string | ?number,
  columnFill?: ?string | ?number,
  columnGap?: ?string | ?number,
  columnRule?: ?string | ?number,
  columnRuleColor?: ?string | ?number,
  columnRuleStyle?: ?string | ?number,
  columnRuleWidth?: ?string | ?number,
  columnSpan?: ?string | ?number,
  columnWidth?: ?string | ?number,
  columns?: ?string | ?number,
  content?: ?string | ?number,
  counterIncrement?: ?string | ?number,
  counterReset?: ?string | ?number,
  cursor?: ?string | ?number,
  deg?: ?string | ?number,
  direction?: ?string | ?number,
  display?: ?string | ?number,
  dpcm?: ?string | ?number,
  dpi?: ?string | ?number,
  dppx?: ?string | ?number,
  em?: ?string | ?number,
  emptyCells?: ?string | ?number,
  ex?: ?string | ?number,
  filter?: ?string | ?number,
  flex?: ?string | ?number,
  flexBasis?: ?string | ?number,
  flexDirection?: ?string | ?number,
  flexFlow?: ?string | ?number,
  flexGrow?: ?string | ?number,
  flexShrink?: ?string | ?number,
  flexWrap?: ?string | ?number,
  float?: ?string | ?number,
  font?: ?string | ?number,
  fontFamily?: ?string | ?number,
  fontFeatureSettings?: ?string | ?number,
  fontKerning?: ?string | ?number,
  fontLanguageOverride?: ?string | ?number,
  fontSize?: ?string | ?number,
  fontSizeAdjust?: ?string | ?number,
  fontStretch?: ?string | ?number,
  fontStyle?: ?string | ?number,
  fontSynthesis?: ?string | ?number,
  fontVariant?: ?string | ?number,
  fontVariantAlternates?: ?string | ?number,
  fontVariantCaps?: ?string | ?number,
  fontVariantEastAsian?: ?string | ?number,
  fontVariantLigatures?: ?string | ?number,
  fontVariantNumeric?: ?string | ?number,
  fontVariantPosition?: ?string | ?number,
  fontWeight?: ?string | ?number,
  fr?: ?string | ?number,
  grad?: ?string | ?number,
  grid?: ?string | ?number,
  gridArea?: ?string | ?number,
  gridAutoColumns?: ?string | ?number,
  gridAutoFlow?: ?string | ?number,
  gridAutoRows?: ?string | ?number,
  gridColumn?: ?string | ?number,
  gridColumnEnd?: ?string | ?number,
  gridColumnGap?: ?string | ?number,
  gridColumnStart?: ?string | ?number,
  gridGap?: ?string | ?number,
  gridRow?: ?string | ?number,
  gridRowEnd?: ?string | ?number,
  gridRowGap?: ?string | ?number,
  gridRowStart?: ?string | ?number,
  gridTemplate?: ?string | ?number,
  gridTemplateAreas?: ?string | ?number,
  gridTemplateColumns?: ?string | ?number,
  gridTemplateRows?: ?string | ?number,
  hangingPunctuation?: ?string | ?number,
  height?: ?string | ?number,
  hyphens?: ?string | ?number,
  hz?: ?string | ?number,
  imageOrientation?: ?string | ?number,
  imageRendering?: ?string | ?number,
  imageResolution?: ?string | ?number,
  in?: ?string | ?number,
  inherit?: ?string | ?number,
  initial?: ?string | ?number,
  inlineSize?: ?string | ?number,
  isolation?: ?string | ?number,
  justifyContent?: ?string | ?number,
  khz?: ?string | ?number,
  left?: ?string | ?number,
  letterSpacing?: ?string | ?number,
  lineBreak?: ?string | ?number,
  lineHeight?: ?string | ?number,
  listStyle?: ?string | ?number,
  listStyleImage?: ?string | ?number,
  listStylePosition?: ?string | ?number,
  listStyleType?: ?string | ?number,
  margin?: ?string | ?number,
  marginBlockEnd?: ?string | ?number,
  marginBlockStart?: ?string | ?number,
  marginBottom?: ?string | ?number,
  marginInlineEnd?: ?string | ?number,
  marginInlineStart?: ?string | ?number,
  marginLeft?: ?string | ?number,
  marginRight?: ?string | ?number,
  marginTop?: ?string | ?number,
  mask?: ?string | ?number,
  maskClip?: ?string | ?number,
  maskComposite?: ?string | ?number,
  maskImage?: ?string | ?number,
  maskMode?: ?string | ?number,
  maskOrigin?: ?string | ?number,
  maskPosition?: ?string | ?number,
  maskRepeat?: ?string | ?number,
  maskSize?: ?string | ?number,
  maskType?: ?string | ?number,
  maxHeight?: ?string | ?number,
  maxWidth?: ?string | ?number,
  minBlockSize?: ?string | ?number,
  minHeight?: ?string | ?number,
  minInlineSize?: ?string | ?number,
  minWidth?: ?string | ?number,
  mixBlendMode?: ?string | ?number,
  mm?: ?string | ?number,
  ms?: ?string | ?number,
  objectFit?: ?string | ?number,
  objectPosition?: ?string | ?number,
  offsetBlockEnd?: ?string | ?number,
  offsetBlockStart?: ?string | ?number,
  offsetInlineEnd?: ?string | ?number,
  offsetInlineStart?: ?string | ?number,
  opacity?: ?string | ?number,
  order?: ?string | ?number,
  orphans?: ?string | ?number,
  outline?: ?string | ?number,
  outlineColor?: ?string | ?number,
  outlineOffset?: ?string | ?number,
  outlineStyle?: ?string | ?number,
  outlineWidth?: ?string | ?number,
  overflow?: ?string | ?number,
  overflowWrap?: ?string | ?number,
  overflowX?: ?string | ?number,
  overflowY?: ?string | ?number,
  padding?: ?string | ?number,
  paddingBlockEnd?: ?string | ?number,
  paddingBlockStart?: ?string | ?number,
  paddingBottom?: ?string | ?number,
  paddingInlineEnd?: ?string | ?number,
  paddingInlineStart?: ?string | ?number,
  paddingLeft?: ?string | ?number,
  paddingRight?: ?string | ?number,
  paddingTop?: ?string | ?number,
  pageBreakAfter?: ?string | ?number,
  pageBreakBefore?: ?string | ?number,
  pageBreakInside?: ?string | ?number,
  pc?: ?string | ?number,
  perspective?: ?string | ?number,
  perspectiveOrigin?: ?string | ?number,
  pointerEvents?: ?string | ?number,
  position?: ?string | ?number,
  pt?: ?string | ?number,
  px?: ?string | ?number,
  q?: ?string | ?number,
  quotes?: ?string | ?number,
  rad?: ?string | ?number,
  rem?: ?string | ?number,
  resize?: ?string | ?number,
  revert?: ?string | ?number,
  right?: ?string | ?number,
  rubyAlign?: ?string | ?number,
  rubyMerge?: ?string | ?number,
  rubyPosition?: ?string | ?number,
  s?: ?string | ?number,
  scrollBehavior?: ?string | ?number,
  scrollSnapCoordinate?: ?string | ?number,
  scrollSnapDestination?: ?string | ?number,
  scrollSnapType?: ?string | ?number,
  shapeImageThreshold?: ?string | ?number,
  shapeMargin?: ?string | ?number,
  shapeOutside?: ?string | ?number,
  tabSize?: ?string | ?number,
  tableLayout?: ?string | ?number,
  textAlign?: ?string | ?number,
  textAlignLast?: ?string | ?number,
  textCombineUpright?: ?string | ?number,
  textDecoration?: ?string | ?number,
  textDecorationColor?: ?string | ?number,
  textDecorationLine?: ?string | ?number,
  textDecorationStyle?: ?string | ?number,
  textEmphasis?: ?string | ?number,
  textEmphasisColor?: ?string | ?number,
  textEmphasisPosition?: ?string | ?number,
  textEmphasisStyle?: ?string | ?number,
  textIndent?: ?string | ?number,
  textJustify?: ?string | ?number,
  textOrientation?: ?string | ?number,
  textOverflow?: ?string | ?number,
  textRendering?: ?string | ?number,
  textShadow?: ?string | ?number,
  textTransform?: ?string | ?number,
  textUnderlinePosition?: ?string | ?number,
  top?: ?string | ?number,
  touchAction?: ?string | ?number,
  transform?: ?string | ?number,
  transformBox?: ?string | ?number,
  transformOrigin?: ?string | ?number,
  transformStyle?: ?string | ?number,
  transition?: ?string | ?number,
  transitionDelay?: ?string | ?number,
  transitionDuration?: ?string | ?number,
  transitionProperty?: ?string | ?number,
  transitionTimingFunction?: ?string | ?number,
  turn?: ?string | ?number,
  unicodeBidi?: ?string | ?number,
  unset?: ?string | ?number,
  verticalAlign?: ?string | ?number,
  vh?: ?string | ?number,
  visibility?: ?string | ?number,
  vmax?: ?string | ?number,
  vmin?: ?string | ?number,
  vw?: ?string | ?number,
  whiteSpace?: ?string | ?number,
  widows?: ?string | ?number,
  width?: ?string | ?number,
  willChange?: ?string | ?number,
  wordBreak?: ?string | ?number,
  wordSpacing?: ?string | ?number,
  wordWrap?: ?string | ?number,
  writingMode?: ?string | ?number,
  zIndex?: ?string | ?number,
  overflowScrolling?: ?string | ?number,
  userSelect?: ?string | ?number,
}
