// react-hot-loader/dist/webpack.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var sourceMap = _interopDefault(require('source-map'));
var fs = _interopDefault(require('fs'));
var path = _interopDefault(require('path'));
var loaderUtils = _interopDefault(require('loader-utils'));

var SourceMapGenerator = sourceMap.SourceMapGenerator;

function makeIdentitySourceMap(content, resourcePath) {
  var map = new SourceMapGenerator();
  map.setSourceContent(resourcePath, content);

  content.split('\n').forEach(function (line, index) {
    map.addMapping({
      source: resourcePath,
      original: {
        line: index + 1,
        column: 0
      },
      generated: {
        line: index + 1,
        column: 0
      }
    });
  });

  return map.toJSON();
}

var makeIdentitySourceMap_1 = makeIdentitySourceMap;

var injectionStart = {
  '16.6': ['if (child.tag === Fragment ? element.type === REACT_FRAGMENT_TYPE : child.elementType === element.type)', 'if (child.tag === Fragment ? element.type === REACT_FRAGMENT_TYPE : hotCompareElements(child.elementType, element.type, updateChild(child)))'],
  '16.6-compact': ['if(child.tag===Fragment?element.type===REACT_FRAGMENT_TYPE:child.elementType===element.type)', 'if(child.tag===Fragment?element.type===REACT_FRAGMENT_TYPE:hotCompareElements(child.elementType,element.type, updateChild(child)))'],
  '16.4': ['if (child.tag === Fragment ? element.type === REACT_FRAGMENT_TYPE : child.type === element.type) {', 'if (child.tag === Fragment ? element.type === REACT_FRAGMENT_TYPE : hotCompareElements(child.type, element.type, updateChild(child))) {'],
  '16.4-compact': ['if(child.tag===Fragment?element.type===REACT_FRAGMENT_TYPE:child.type===element.type)', 'if(child.tag===Fragment?element.type===REACT_FRAGMENT_TYPE:hotCompareElements(child.type,element.type, updateChild(child)))']
};

var additional = {
  '16.6-update': ['if (current$$1 !== null && current$$1.elementType === element.type) {', 'if (current$$1 !== null && hotCompareElements(current$$1.elementType, element.type, updateChild(current$$1))) {'],
  '16.6-update-compact': ['if(current$$1!==null&&current$$1.elementType===element.type)', 'if(current$$1!==null&&hotCompareElements(current$$1.elementType,element.type,updateChild(current$$1)))'],
  '16.4-update': ['if (current !== null && current.type === element.type) {', 'if (current !== null && hotCompareElements(current.type, element.type, updateChild(current))) {'],
  '16.4-update-compact': ['if (current!== null&&current.type===element.type)', 'if (current!== null&&hotCompareElements(current.type,element.type,updateChild(current)))']
};

var defaultEnd = ['var ReactDOM = {', 'var updateChild = function (child) { return function (newType) {\n       child.type = newType; \n       if (child.alternate) {\n        child.alternate.type = newType;\n       }\n    }};\n    var hotCompareElements = function (oldType, newType) { return oldType === newType };\n    var ReactDOM = {\n      setHotElementComparator: function (newComparator) { hotCompareElements = newComparator }, \n  '];

var defaultEndCompact = ['var ReactDOM={', 'var updateChild = function (child) { return function (newType) {\n       child.type = newType; \n       if (child.alternate) {\n        child.alternate.type = newType;\n       }\n    }};\n    var hotCompareElements = function (oldType, newType) { return oldType === newType };\n    var ReactDOM = {\n      setHotElementComparator: function (newComparator) { hotCompareElements = newComparator }, \n  '];

var injectionEnd = {
  '16.6': defaultEnd,
  '16.4': defaultEnd,
  '16.6-compact': defaultEndCompact,
  '16.4-compact': defaultEndCompact
};

function additionalTransform(source) {
  for (var key in additional) {
    source = source.split(additional[key][0]).join(additional[key][1]);
  }
  return source;
}

function transform(source) {
  if (source.indexOf('reconcileSingleElement') < 0) {
    // early reject
    return source;
  }
  for (var key in injectionStart) {
    if (source.indexOf(injectionStart[key][0]) > 0 && source.indexOf(injectionEnd[key][0]) > 0) {
      return additionalTransform(source.replace(injectionStart[key][0], injectionStart[key][1]).replace(injectionEnd[key][0], injectionEnd[key][1])) + '/** react is now 🔥 */';
    }
  }
  return source;
}

var patch = transform;

var SourceNode = sourceMap.SourceNode,
    SourceMapConsumer = sourceMap.SourceMapConsumer;





var tagCommonJSExportsSource = null;

function transform$1(source, map) {
  var _this = this;

  var callback = this.async();
  if (source && source.types && source.types.IfStatement) {
    throw new Error('React Hot Loader: You are erroneously trying to use a Webpack loader ' + 'as a Babel plugin. Replace "react-hot-loader/webpack" with ' + '"react-hot-loader/babel" in the "plugins" section of your .babelrc file. ' + 'While we recommend the above, if you prefer not to use Babel, ' + 'you may remove "react-hot-loader/webpack" from the "plugins" section of ' + 'your .babelrc file altogether, and instead add "react-hot-loader/webpack" ' + 'to the "loaders" section of your Webpack configuration.');
  }

  if (this.cacheable) {
    this.cacheable();
  }

  var options = Object.assign({ withPatch: true }, loaderUtils.getOptions(this));
  if (options.withPatch) {
    source = patch(source);
  }
  if (source.indexOf('reactHotLoader.register') > 0) {
    return callback(null, source, map);
  }
  // This is a Webpack loader, but the user put it in the Babel config.

  // Read the helper once.
  if (!tagCommonJSExportsSource) {
    tagCommonJSExportsSource = fs.readFileSync(path.join(__dirname, 'webpackTagCommonJSExports.js'), 'utf8')
    // Babel inserts these.
    // Ideally we'd opt out for one file but this is simpler.
    .replace(/['"]use strict['"];/, '')
    // eslint comments don't need to end up in the output
    .replace(/\/\/ eslint-disable-line .*\n/g, '\n').replace(/\/\* global.*\*\//, '').split(/\n\s*/).join(' ');
  }

  // Parameterize the helper with the current filename.
  var separator = '\n\n;';
  var appendText = tagCommonJSExportsSource.replace(/__FILENAME__/g, JSON.stringify(this.resourcePath));

  if (this.sourceMap === false) {
    return callback(null, [source, appendText].join(separator));
  }

  if (!map) {
    map = makeIdentitySourceMap_1(source, this.resourcePath); // eslint-disable-line no-param-reassign
  }
  var sourceMapConsumer = new SourceMapConsumer(map);
  sourceMapConsumer.then(function (consumedMap) {
    var node = new SourceNode(null, null, null, [SourceNode.fromStringWithSourceMap(source, consumedMap), new SourceNode(null, null, _this.resourcePath, appendText)]).join(separator);
    var result = node.toStringWithSourceMap();
    callback(null, result.code, result.map.toJSON() || undefined);
  });
}

var webpack = transform$1;

exports.default = webpack;
