// @superset-ui/color/lib/CategoricalColorNamespace.js
"use strict";

exports.__esModule = true;
exports.getNamespace = getNamespace;
exports.getColor = getColor;
exports.getScale = getScale;
exports.DEFAULT_NAMESPACE = exports.default = void 0;

var _CategoricalColorScale = _interopRequireDefault(require("./CategoricalColorScale"));

var _CategoricalSchemeRegistrySingleton = _interopRequireDefault(require("./CategoricalSchemeRegistrySingleton"));

var _stringifyAndTrim = _interopRequireDefault(require("./stringifyAndTrim"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class CategoricalColorNamespace {
  constructor(name) {
    _defineProperty(this, "name", void 0);

    _defineProperty(this, "forcedItems", void 0);

    _defineProperty(this, "scales", void 0);

    this.name = name;
    this.scales = {};
    this.forcedItems = {};
  }

  getScale(schemeId) {
    var _ref, _scheme$colors;

    const id = (_ref = schemeId != null ? schemeId : (0, _CategoricalSchemeRegistrySingleton.default)().getDefaultKey()) != null ? _ref : '';
    const scale = this.scales[id];

    if (scale) {
      return scale;
    }

    const scheme = (0, _CategoricalSchemeRegistrySingleton.default)().get(id);
    const newScale = new _CategoricalColorScale.default((_scheme$colors = scheme == null ? void 0 : scheme.colors) != null ? _scheme$colors : [], this.forcedItems);
    this.scales[id] = newScale;
    return newScale;
  }
  /**
   * Enforce specific color for given value
   * This will apply across all color scales
   * in this namespace.
   * @param {*} value value
   * @param {*} forcedColor color
   */


  setColor(value, forcedColor) {
    this.forcedItems[(0, _stringifyAndTrim.default)(value)] = forcedColor;
    return this;
  }

}

exports.default = CategoricalColorNamespace;
const namespaces = {};
const DEFAULT_NAMESPACE = 'GLOBAL';
exports.DEFAULT_NAMESPACE = DEFAULT_NAMESPACE;

function getNamespace(name = DEFAULT_NAMESPACE) {
  const instance = namespaces[name];

  if (instance) {
    return instance;
  }

  const newInstance = new CategoricalColorNamespace(name);
  namespaces[name] = newInstance;
  return newInstance;
}

function getColor(value, schemeId, namespace) {
  return getNamespace(namespace).getScale(schemeId).getColor(value);
}

function getScale(scheme, namespace) {
  return getNamespace(namespace).getScale(scheme);
}
