// @superset-ui/legacy-preset-chart-nvd3/lib/TimePivot/controlPanel.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _translation = require("@superset-ui/translation");

var _chartControls = require("@superset-ui/chart-controls");

var _NVD3Controls = require("../NVD3Controls");

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var _default = {
  requiresTime: true,
  controlPanelSections: [{
    label: (0, _translation.t)('Query'),
    expanded: true,
    controlSetRows: [['metric'], ['adhoc_filters'], [{
      name: 'freq',
      config: {
        type: 'SelectControl',
        label: (0, _translation.t)('Frequency'),
        default: 'W-MON',
        freeForm: true,
        clearable: false,
        choices: [['AS', 'Year (freq=AS)'], ['52W-MON', '52 weeks starting Monday (freq=52W-MON)'], ['W-SUN', '1 week starting Sunday (freq=W-SUN)'], ['W-MON', '1 week starting Monday (freq=W-MON)'], ['D', 'Day (freq=D)'], ['4W-MON', '4 weeks (freq=4W-MON)']],
        description: (0, _translation.t)("The periodicity over which to pivot time. Users can provide\n            \"Pandas\" offset alias.\n            Click on the info bubble for more details on accepted \"freq\" expressions."),
        tooltipOnClick: () => {
          window.open('https://pandas.pydata.org/pandas-docs/stable/timeseries.html#offset-aliases');
        }
      }
    }]]
  }, {
    label: (0, _translation.t)('Chart Options'),
    expanded: true,
    controlSetRows: [[_NVD3Controls.showLegend, _NVD3Controls.lineInterpolation], ['color_picker', null]]
  }, {
    label: (0, _translation.t)('X Axis'),
    expanded: true,
    controlSetRows: [[_NVD3Controls.xAxisLabel, _NVD3Controls.bottomMargin], [_NVD3Controls.xAxisShowMinmax, {
      name: _NVD3Controls.xAxisFormat.name,
      config: _extends({}, _NVD3Controls.xAxisFormat.config, {
        default: 'SMART_NUMBER',
        choices: _chartControls.D3_FORMAT_OPTIONS
      })
    }]]
  }, {
    label: (0, _translation.t)('Y Axis'),
    expanded: true,
    controlSetRows: [[_NVD3Controls.yAxisLabel, _NVD3Controls.leftMargin], [_NVD3Controls.yAxisShowMinmax, _NVD3Controls.yLogScale], ['y_axis_format', _NVD3Controls.yAxisBounds]]
  }],
  controlOverrides: {
    metric: {
      clearable: false
    }
  },
  sectionOverrides: {
    druidTimeSeries: {
      controlSetRows: [['granularity', 'druid_time_origin'], ['time_range']]
    },
    sqlaTimeSeries: {
      controlSetRows: [['granularity_sqla', 'time_grain_sqla'], ['time_range']]
    }
  }
};
exports.default = _default;
